--- Devel-NYTProf-2.09/Makefile.PL.orig	2009-03-06 11:57:18.000000000 -0500
+++ Devel-NYTProf-2.09/Makefile.PL	2009-03-06 11:57:43.000000000 -0500
@@ -49,10 +49,11 @@
 
 # --- Discover how much of stdio is implemented
 
-my $cpp = $Config{cpp} || do {
-    warn "Warning: cpp not found in your perl config.  Falling back to 'cat'\n";
-    'cat';
-};
+#my $cpp = $Config{cpp} || do {
+#    warn "Warning: cpp not found in your perl config.  Falling back to 'cat'\n";
+#    'cat';
+#};
+my $cpp = 'cat';
 
 print "Looking for header files and functions...\n";
 my $INCLUDE;
