diff -ur XML-Parser-2.41-orig/Expat/Makefile.PL XML-Parser-2.41/Expat/Makefile.PL
--- XML-Parser-2.41-orig/Expat/Makefile.PL	2011-06-03 01:35:32.000000000 -0400
+++ XML-Parser-2.41/Expat/Makefile.PL	2011-09-12 16:38:33.000000000 -0400
@@ -1,6 +1,6 @@
 use ExtUtils::MakeMaker;
 use Config;
-use English;
+#use English;
 
 my $libs = "-lexpat";
 my @extras = ();
diff -ur XML-Parser-2.41-orig/Makefile.PL XML-Parser-2.41/Makefile.PL
--- XML-Parser-2.41-orig/Makefile.PL	2011-09-12 16:36:32.000000000 -0400
+++ XML-Parser-2.41/Makefile.PL	2011-09-12 16:38:08.000000000 -0400
@@ -2,7 +2,6 @@
 use 5.004005; #Devel::CheckLib
 use ExtUtils::MakeMaker;
 use lib qw(inc);
-use Devel::CheckLib;
 
 use Config;
 
@@ -28,43 +27,6 @@
 
 @ARGV = @replacement_args;
 
-unless (
-  check_lib(     # fill in what you prompted the user for here
-    lib => [qw(expat)],
-    header => ['expat.h'],
-    incpath => $expat_incpath,
-    ($expat_libpath?
-    (libpath => $expat_libpath):()),
-  )) {
-    warn <<'Expat_Not_Installed;';
-
-Expat must be installed prior to building XML::Parser and I can't find
-it in the standard library directories. Install 'expat-devel' package with your
-OS package manager. See 'README'.
-
-Or you can download expat from:
-
-http://sourceforge.net/projects/expat/
-
-If expat is installed, but in a non-standard directory, then use the
-following options to Makefile.PL:
-
-    EXPATLIBPATH=...  To set the directory in which to find libexpat
-
-    EXPATINCPATH=...  To set the directory in which to find expat.h
-
-For example:
-
-    perl Makefile.PL EXPATLIBPATH=/home/me/lib EXPATINCPATH=/home/me/include
-
-Note that if you build against a shareable library in a non-standard location
-you may (on some platforms) also have to set your LD_LIBRARY_PATH environment
-variable at run time for perl to find the library.
-
-Expat_Not_Installed;
-    #exit;
-}
-
 if (not $expat_libpath and $] >= 5.006001 and $^O ne 'MSWin32') {
   require ExtUtils::Liblist;		# Buggy before this
   ($expat_libpath) = ExtUtils::Liblist->ext('-lexpat');
