From c4002d37e84ce784ab18aa32febd86b22c460826 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 24 Mar 2007 18:19:49 -0400
Subject: [PATCH] coding-style: remove unnecessary {braces} around one-line code blocks

---
 fs/unionfs/dentry.c |    9 +++------
 1 files changed, 3 insertions(+), 6 deletions(-)

diff --git a/fs/unionfs/dentry.c b/fs/unionfs/dentry.c
index 17c0b27..ab3f9e6 100644
--- a/fs/unionfs/dentry.c
+++ b/fs/unionfs/dentry.c
@@ -196,9 +196,8 @@ int __unionfs_d_revalidate_chain(struct dentry *dentry, struct nameidata *nd)
 		dtmp = dtmp->d_parent;
 		dgen = atomic_read(&UNIONFS_D(dtmp)->generation);
 	}
-	if (chain_len == 0) {
+	if (chain_len == 0)
 		goto out_this;	/* shortcut if parents are OK */
-	}
 
 	/*
 	 * Allocate array of dentries to reval.  We could use linked lists,
@@ -242,9 +241,8 @@ int __unionfs_d_revalidate_chain(struct dentry *dentry, struct nameidata *nd)
 		}
 		unionfs_unlock_dentry(chain[i]);
 
-		if (!valid) {
+		if (!valid)
 			goto out_free;
-		}
 	}
 
 
@@ -264,9 +262,8 @@ out_this:
 out_free:
 	/* unlock/dput all dentries in chain and return status */
 	if (chain_len > 0) {
-		for (i=0; i<chain_len; i++) {
+		for (i=0; i<chain_len; i++)
 			dput(chain[i]);
-		}
 		kfree(chain);
 	}
 out:
-- 
1.5.4.3

