From 40627d1d7b4bb0acb4b169c3264a45ec6e08f141 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 24 Mar 2007 19:32:42 -0400
Subject: [PATCH] coding-style: remove one more {braces} around one-statement code block

---
 fs/unionfs/subr.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/subr.c b/fs/unionfs/subr.c
index 4f1b355..f77101b 100644
--- a/fs/unionfs/subr.c
+++ b/fs/unionfs/subr.c
@@ -77,12 +77,11 @@ int create_whiteout(struct dentry *dentry, int start)
 		}
 
 		hidden_dir_dentry = lock_parent(hidden_wh_dentry);
-		if (!(err = is_robranch_super(dentry->d_sb, bindex))) {
+		if (!(err = is_robranch_super(dentry->d_sb, bindex)))
 			err = vfs_create(hidden_dir_dentry->d_inode,
 					 hidden_wh_dentry,
 					 ~current->fs->umask & S_IRWXUGO,
 					 NULL);
-		}
 		unlock_dir(hidden_dir_dentry);
 		dput(hidden_wh_dentry);
 
-- 
1.5.4.3

