From cc9787ec4a9a124576347aec57dd867d6a3a7114 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 24 Mar 2007 23:01:13 -0400
Subject: [PATCH] cleanup: one more trailing whitespace removed.

---
 fs/unionfs/union.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/union.h b/fs/unionfs/union.h
index 6a03018..f634d17 100644
--- a/fs/unionfs/union.h
+++ b/fs/unionfs/union.h
@@ -462,7 +462,7 @@ static inline struct vfsmount *unionfs_mntget(struct dentry *dentry,
 					      int bindex)
 {
 	struct vfsmount *mnt;
- 
+
 	if (!dentry) {
 		if (bindex < 0)
 			return NULL;
-- 
1.5.4.3

