From f30d606ef9cdbb32baade01738689c887eb71387 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Fri, 4 May 2007 14:44:02 -0400
Subject: [PATCH] optimize branch overlapping test a bit

---
 fs/unionfs/main.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/main.c b/fs/unionfs/main.c
index 7723190..ffcd040 100644
--- a/fs/unionfs/main.c
+++ b/fs/unionfs/main.c
@@ -366,10 +366,9 @@ static int parse_dirs_option(struct super_block *sb, struct unionfs_dentry_info
 	 * branch-overlapping test.
 	 */
 	for (i = 0; i < branches; i++) {
+		dent1 = hidden_root_info->lower_paths[i].dentry;
 		for (j = i + 1; j < branches; j++) {
-			dent1 = hidden_root_info->lower_paths[i].dentry;
 			dent2 = hidden_root_info->lower_paths[j].dentry;
-
 			if (is_branch_overlap(dent1, dent2)) {
 				printk(KERN_WARNING "unionfs: branches %d and "
 				       "%d overlap\n", i, j);
-- 
1.5.4.3

