From 5f9e5bffb185dff9c1113d5ac06aa106abbc31de Mon Sep 17 00:00:00 2001
From: Yiannis Pericleous <yiannos@louie.fsl.cs.sunysb.edu>
Date: Sat, 28 Apr 2007 22:06:46 -0400
Subject: [PATCH] cleanup: removed a trailing whitespace

---
 fs/unionfs/super.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/super.c b/fs/unionfs/super.c
index 901840e..86a0b01 100644
--- a/fs/unionfs/super.c
+++ b/fs/unionfs/super.c
@@ -725,7 +725,8 @@ out_no_change:
 	i = atomic_inc_return(&UNIONFS_SB(sb)->generation);
 	atomic_set(&UNIONFS_D(sb->s_root)->generation, i);
 	atomic_set(&UNIONFS_I(sb->s_root->d_inode)->generation, i);
-	printk("unionfs: new generation number %d\n", i);
+	if (!(*flags & MS_SILENT))
+		printk("unionfs: new generation number %d\n", i);
 	err = 0;		/* reset to success */
 
 	/*
-- 
1.5.4.3

