From a3312294d39c33a9624886eaa0d73d8fc8266377 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 5 May 2007 19:10:00 -0400
Subject: [PATCH] mmap: trivial sendfile support

---
 fs/unionfs/file.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/fs/unionfs/file.c b/fs/unionfs/file.c
index 0650cb9..c4deba6 100644
--- a/fs/unionfs/file.c
+++ b/fs/unionfs/file.c
@@ -127,4 +127,5 @@ struct file_operations unionfs_main_fops = {
 	.flush		= unionfs_flush,
 	.release	= unionfs_file_release,
 	.fsync		= file_fsync,
+	.sendfile	= generic_file_sendfile,
 };
-- 
1.5.4.3

