From c7c2113a7a452cd21a947261243f5f06bd27d7c2 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 19 May 2007 21:11:22 -0400
Subject: [PATCH] Add Erez Zadok to maintainers list

---
 MAINTAINERS |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/MAINTAINERS b/MAINTAINERS
index 1992b5e..161652b 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -3594,6 +3594,8 @@ W:	http://www.kernel.dk
 S:	Maintained
 
 UNIONFS
+P:	Erez Zadok
+M:	ezk@cs.sunysb.edu
 P:	Josef "Jeff" Sipek
 M:	jsipek@cs.sunysb.edu
 L:	unionfs@filesystems.org
-- 
1.5.4.3

