From 3390b194f951bcd8653fc91a5abda90e7a4dcb5f Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@bigvaio.(none)>
Date: Tue, 22 May 2007 23:53:25 -0400
Subject: [PATCH] documentation: comment to ODF developers to address

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/main.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/fs/unionfs/main.c b/fs/unionfs/main.c
index f83ea00..decefef 100644
--- a/fs/unionfs/main.c
+++ b/fs/unionfs/main.c
@@ -215,6 +215,7 @@ void unionfs_reinterpose(struct dentry *dentry)
  */
 int check_branch(struct nameidata *nd)
 {
+	/* XXX: remove in ODF code -- stacking unions allowed there */
 	if (!strcmp(nd->dentry->d_sb->s_type->name, "unionfs"))
 		return -EINVAL;
 	if (!nd->dentry->d_inode)
-- 
1.5.4.3

