From 0464bc902d2127c0b06ebaeab373dd54b8906d66 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Fri, 25 May 2007 22:36:13 -0400
Subject: [PATCH] Unionfs: allow users to override Makefile options in fs/unionfs/local.mk

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/Makefile |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/fs/unionfs/Makefile b/fs/unionfs/Makefile
index 883427c..7c18a3c 100644
--- a/fs/unionfs/Makefile
+++ b/fs/unionfs/Makefile
@@ -8,8 +8,12 @@ unionfs-y := subr.o dentry.o file.o inode.o main.o super.o \
 unionfs-$(CONFIG_UNION_FS_XATTR) += xattr.o
 
 # If you want debugging output, please uncomment the following line
+# or put your options in a separate file in linux-x.y.z/fs/unionfs/local.mk
 #CONFIG_UNIONFS_DEBUG=y
 
+# Allow users to override debug options in a separate file
+-include fs/unionfs/local.mk
+
 ifeq ($(CONFIG_UNIONFS_DEBUG),y)
 unionfs-y += debug.o
 EXTRA_CFLAGS += -DUNIONFS_DEBUG=1
-- 
1.5.4.3

