From 819739f3468013b5b5570b0cb3c16407e9221cbd Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Mon, 16 Jul 2007 23:42:55 -0400
Subject: [PATCH] Unionfs: cache-coherency: update times of root inode after branch management

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/super.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/fs/unionfs/super.c b/fs/unionfs/super.c
index 9b6f501..c168b23 100644
--- a/fs/unionfs/super.c
+++ b/fs/unionfs/super.c
@@ -764,6 +764,8 @@ out_no_change:
 	atomic_set(&UNIONFS_I(sb->s_root->d_inode)->generation, i);
 	if (!(*flags & MS_SILENT))
 		printk("unionfs: new generation number %d\n", i);
+	/* finally, update the root dentry's times */
+	unionfs_copy_attr_times(sb->s_root->d_inode);
 	err = 0;		/* reset to success */
 
 	/*
-- 
1.5.4.3

