From beedca83d9b6787984cc5da74dc75306e9535d09 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Tue, 17 Jul 2007 03:18:25 -0400
Subject: [PATCH] Unionfs: cleanup: break line longer than 80 chars

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/commonfops.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/commonfops.c b/fs/unionfs/commonfops.c
index 42678a6..f6f7dea 100644
--- a/fs/unionfs/commonfops.c
+++ b/fs/unionfs/commonfops.c
@@ -368,7 +368,8 @@ int unionfs_file_revalidate(struct file *file, int willwrite)
 			err = open_highest_file(file, willwrite);
 			if (err)
 				goto out;
-			new_brid = UNIONFS_F(file)->saved_branch_ids[fbstart(file)];
+			new_brid = UNIONFS_F(file)->
+			  saved_branch_ids[fbstart(file)];
 			if (new_brid != orig_brid) {
 				/*
 				 * If we re-opened the file on a different
-- 
1.5.4.3

