From 83e13ddf0332a2350d16ee299ad92af286fc0594 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Wed, 22 Aug 2007 18:09:24 -0400
Subject: [PATCH] Unionfs: release 2.1.2

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/Makefile b/fs/unionfs/Makefile
index c8b72e8..c2444ce 100644
--- a/fs/unionfs/Makefile
+++ b/fs/unionfs/Makefile
@@ -1,4 +1,4 @@
-UNIONFS_VERSION="2.1.1 (for 2.6.22.3)"
+UNIONFS_VERSION="2.1.2 (for 2.6.22.4)"
 
 EXTRA_CFLAGS += -DUNIONFS_VERSION=\"$(UNIONFS_VERSION)\"
 
-- 
1.5.4.3

