From 36f0aae8d3015053e7f4fd98822e72a11b47eefe Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Thu, 6 Sep 2007 00:42:56 -0400
Subject: [PATCH] Unionfs: cache-coherency fix for is_newer_lower

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/dentry.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/dentry.c b/fs/unionfs/dentry.c
index 1891df9..e46a53e 100644
--- a/fs/unionfs/dentry.c
+++ b/fs/unionfs/dentry.c
@@ -184,7 +184,7 @@ out:
 
 /*
  * Determine if the lower inode objects have changed from below the unionfs
- * inode.  Return 1 if changed, 0 otherwise.
+ * inode.  Return true if changed, false otherwise.
  */
 bool is_newer_lower(const struct dentry *dentry)
 {
@@ -226,7 +226,7 @@ bool is_newer_lower(const struct dentry *dentry)
 			return true; /* ctime changed! */
 		}
 	}
-	return true;		/* default: lower is not newer */
+	return false;		/* default: lower is not newer */
 }
 
 /*
-- 
1.5.4.3

