From 59a9e722390cd79f258b3944538d73877eb78f86 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sat, 29 Sep 2007 20:53:26 -0400
Subject: [PATCH] Unionfs: use UNIONFS_NAME macro

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/main.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/main.c b/fs/unionfs/main.c
index 72438fb..615617a 100644
--- a/fs/unionfs/main.c
+++ b/fs/unionfs/main.c
@@ -227,7 +227,7 @@ void unionfs_reinterpose(struct dentry *dentry)
 int check_branch(struct nameidata *nd)
 {
 	/* XXX: remove in ODF code -- stacking unions allowed there */
-	if (!strcmp(nd->dentry->d_sb->s_type->name, "unionfs"))
+	if (!strcmp(nd->dentry->d_sb->s_type->name, UNIONFS_NAME))
 		return -EINVAL;
 	if (!nd->dentry->d_inode)
 		return -ENOENT;
@@ -722,7 +722,7 @@ static int unionfs_get_sb(struct file_system_type *fs_type,
 
 static struct file_system_type unionfs_fs_type = {
 	.owner		= THIS_MODULE,
-	.name		= "unionfs",
+	.name		= UNIONFS_NAME,
 	.get_sb		= unionfs_get_sb,
 	.kill_sb	= generic_shutdown_super,
 	.fs_flags	= FS_REVAL_DOT,
-- 
1.5.4.3

