From bd81e53b8368dc80399d70a5764e0f588f5e7b61 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sat, 29 Sep 2007 01:36:57 -0400
Subject: [PATCH] Unionfs: add missing newlines in printk's

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/main.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/main.c b/fs/unionfs/main.c
index 5c7f672..185ddb8 100644
--- a/fs/unionfs/main.c
+++ b/fs/unionfs/main.c
@@ -175,7 +175,7 @@ skip:
 		/* Do nothing. */
 		break;
 	default:
-		printk(KERN_ERR "unionfs: invalid interpose flag passed!");
+		printk(KERN_ERR "unionfs: invalid interpose flag passed!\n");
 		BUG();
 	}
 	goto out;
-- 
1.5.4.3

