From 15855a6748c4a32ea1b2ca2c9c1ab13999bab3dd Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sat, 27 Oct 2007 18:42:58 -0400
Subject: [PATCH] Unionfs: update mmap documentation for unionfs_writepage

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/mmap.c |    6 ------
 1 files changed, 0 insertions(+), 6 deletions(-)

diff --git a/fs/unionfs/mmap.c b/fs/unionfs/mmap.c
index b43557e..852a0cc 100644
--- a/fs/unionfs/mmap.c
+++ b/fs/unionfs/mmap.c
@@ -20,12 +20,6 @@
 #include "union.h"
 
 /*
- * Unionfs doesn't implement ->writepages, which is OK with the VFS and
- * keeps our code simpler and smaller.  Nevertheless, somehow, our own
- * ->writepage must be called so we can sync the upper pages with the lower
- * pages: otherwise data changed at the upper layer won't get written to the
- * lower layer.
- *
  * Some lower file systems (e.g., NFS) expect the VFS to call its writepages
  * only, which in turn will call generic_writepages and invoke each of the
  * lower file system's ->writepage.  NFS in particular uses the
-- 
1.5.4.3

