From c76f7bdf9c5fbc9f43daee9ec0914ca901df4ced Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sun, 25 Nov 2007 18:27:27 -0500
Subject: [PATCH] Unionfs: set lower mnt after mkdir which resulted in copyup

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/inode.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 6ec5f52..a8f3590 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -665,8 +665,10 @@ out:
 
 	kfree(name);
 
-	if (!err)
+	if (!err) {
 		unionfs_copy_attr_times(dentry->d_inode);
+		unionfs_postcopyup_setmnt(dentry);
+	}
 	unionfs_check_inode(parent);
 	unionfs_check_dentry(dentry);
 	unionfs_unlock_dentry(dentry);
-- 
1.5.4.3

