From d2d6413a86c31e7b0cadf8fd8b66f3ecca62afbb Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Tue, 25 Dec 2007 16:55:17 -0500
Subject: [PATCH] Unionfs: update inode times after a successful open

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/commonfops.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/commonfops.c b/fs/unionfs/commonfops.c
index 4077907..b8357a7 100644
--- a/fs/unionfs/commonfops.c
+++ b/fs/unionfs/commonfops.c
@@ -583,10 +583,13 @@ out:
 		kfree(UNIONFS_F(file));
 	}
 out_nofree:
-	unionfs_check_inode(inode);
 	if (!err) {
+		dentry = file->f_path.dentry;
+		unionfs_copy_attr_times(dentry->d_parent->d_inode);
+		unionfs_copy_attr_times(inode);
 		unionfs_check_file(file);
-		unionfs_check_dentry(file->f_path.dentry->d_parent);
+		unionfs_check_dentry(dentry->d_parent);
+		unionfs_check_inode(inode);
 	}
 	unionfs_read_unlock(inode->i_sb);
 	return err;
-- 
1.5.4.3

