From 747c7d9f2af98bcf32a003516848b9a925320512 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Mon, 31 Mar 2008 22:32:42 -0400
Subject: [PATCH] Unionfs: release 2.3.2

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/Makefile b/fs/unionfs/Makefile
index 3ff96a3..acd3746 100644
--- a/fs/unionfs/Makefile
+++ b/fs/unionfs/Makefile
@@ -1,4 +1,4 @@
-UNIONFS_VERSION="2.3.1 (for 2.6.22.19)"
+UNIONFS_VERSION="2.3.2 (for 2.6.22.19)"
 
 EXTRA_CFLAGS += -DUNIONFS_VERSION=\"$(UNIONFS_VERSION)\"
 
-- 
1.5.4.3

