From 6569fe7c23a4fb05adbcf439850e6412fa51fea1 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Tue, 15 Apr 2008 13:26:15 -0400
Subject: [PATCH] Unionfs: stop as soon as first writeable branch is found

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/inode.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 04429a1..2f48b60 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -128,6 +128,8 @@ begin:
 		err = check_for_whiteout(dentry, lower_dentry);
 		if (err)
 			continue;
+		/* if get here, we can write to the branch */
+		break;
 	}
 	/*
 	 * If istart wasn't already branch 0, and we got any error, then try
-- 
1.5.4.3

