From c785fed54b3a8b2a69dbd4481fa2aa95ad0e651d Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sun, 27 Jul 2008 19:15:44 -0400
Subject: [PATCH] Unionfs: update maintainers

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 MAINTAINERS |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/MAINTAINERS b/MAINTAINERS
index 161652b..e1052be 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -3596,10 +3596,9 @@ S:	Maintained
 UNIONFS
 P:	Erez Zadok
 M:	ezk@cs.sunysb.edu
-P:	Josef "Jeff" Sipek
-M:	jsipek@cs.sunysb.edu
 L:	unionfs@filesystems.org
 W:	http://unionfs.filesystems.org
+T:	git git.kernel.org/pub/scm/linux/kernel/git/ezk/unionfs.git
 S:	Maintained
 
 USB ACM DRIVER
-- 
1.5.4.3

