From c2b93d4cfb9748d4d84eb7707218b79bdb68cc72 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Fri, 19 Sep 2008 00:02:56 -0400
Subject: [PATCH] Unionfs: fix off-by-one but in mkdir

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/inode.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 3b1ae5f..41a0b1c 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -508,7 +508,7 @@ static int unionfs_mkdir(struct inode *parent, struct dentry *dentry, int mode)
 		if (err)
 			break;
 
-		for (i = bindex + 1; i < bend; i++) {
+		for (i = bindex + 1; i <= bend; i++) {
 			if (unionfs_lower_dentry_idx(dentry, i)) {
 				dput(unionfs_lower_dentry_idx(dentry, i));
 				unionfs_set_lower_dentry_idx(dentry, i, NULL);
-- 
1.5.4.3

