From 0e032bb222aadf0c2fadef77aa83761cdb26b18b Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Fri, 19 Sep 2008 00:02:56 -0400
Subject: [PATCH] Unionfs: remove unnecessary calls to copy_attr_times

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/inode.c |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 41a0b1c..9667560 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -210,8 +210,6 @@ static struct dentry *unionfs_lookup(struct inode *parent,
 			unionfs_postcopyup_release(dentry);
 		}
 		unionfs_copy_attr_times(dentry->d_inode);
-		/* parent times may have changed */
-		unionfs_copy_attr_times(dentry->d_parent->d_inode);
 	}
 
 	unionfs_check_inode(parent);
-- 
1.5.4.3

