From 0ef8b41be043d68145ce20e7df357d9f94654fc3 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Fri, 19 Sep 2008 00:02:56 -0400
Subject: [PATCH] Unionfs: just warn in put_super instead of BUG-ing

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/super.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/super.c b/fs/unionfs/super.c
index c68b2f8..8e02c38 100644
--- a/fs/unionfs/super.c
+++ b/fs/unionfs/super.c
@@ -114,7 +114,7 @@ static void unionfs_put_super(struct super_block *sb)
 			       bindex, branch_count(sb, bindex));
 			leaks = 1;
 		}
-	BUG_ON(leaks != 0);
+	WARN_ON(leaks != 0);
 
 	/* decrement lower super references */
 	for (bindex = bstart; bindex <= bend; bindex++) {
-- 
1.5.4.3

