--- ../src/utils/common.c.orig	2009-03-23 15:06:28.000000000 +0100
+++ ../src/utils/common.c	2009-07-22 10:31:28.000000000 +0200
@@ -320,7 +320,7 @@ const char * wpa_ssid_txt(const u8 *ssid
 	os_memcpy(ssid_txt, ssid, ssid_len);
 	ssid_txt[ssid_len] = '\0';
 	for (pos = ssid_txt; *pos != '\0'; pos++) {
-		if ((u8) *pos < 32 || (u8) *pos >= 127)
+		if ((u8) *pos < 32)
 			*pos = '_';
 	}
 	return ssid_txt;
