--- ../src/drivers/driver_wext.c.orig	2009-03-23 15:06:28.000000000 +0100
+++ ../src/drivers/driver_wext.c	2009-06-12 02:52:15.000000000 +0200
@@ -1358,6 +1358,37 @@ static void wext_get_scan_custom(struct 
 		hexstr2bin(spos, tmp + res->ie_len, bytes);
 		res->ie = tmp;
 		res->ie_len += bytes;
+
+// 06/09/09 - fm - check for WPS IE in beacon 
+	} else if (clen > 7 && os_strncmp(custom, "wps_ie=", 7) == 0) {
+		char *spos;
+		int bytes;
+		spos = custom + 7;
+		bytes = custom + clen - spos;
+		if (bytes & 1 || bytes == 0)
+			return;
+		bytes /= 2;
+		tmp = os_realloc(res->ie, res->ie_len + bytes);
+		if (tmp == NULL)
+			return;
+		hexstr2bin(spos, tmp + res->ie_len, bytes);
+		res->ie = tmp;
+		res->ie_len += bytes;
+	} else if (clen > 7 && os_strncmp(custom, "wsc_ie=", 7) == 0) {
+		char *spos;
+		int bytes;
+		spos = custom + 7;
+		bytes = custom + clen - spos;
+		if (bytes & 1 || bytes == 0)
+			return;
+		bytes /= 2;
+		tmp = os_realloc(res->ie, res->ie_len + bytes);
+		if (tmp == NULL)
+			return;
+		hexstr2bin(spos, tmp + res->ie_len, bytes);
+		res->ie = tmp;
+		res->ie_len += bytes;
+
 	} else if (clen > 4 && os_strncmp(custom, "tsf=", 4) == 0) {
 		char *spos;
 		int bytes;
