Index: git/Makefile
===================================================================
--- git.orig/Makefile	2008-07-16 13:39:27.000000000 +0100
+++ git/Makefile	2008-07-16 13:49:31.000000000 +0100
@@ -47,8 +47,7 @@
 .SUFFIXES:
 
 all: $(TARGETS)
-	$(MAKE) -C $(BUILDDIR)/ubi-utils
-	$(MAKE) -C $(BUILDDIR)/mkfs.ubifs
+	$(MAKE) -C $(BUILDDIR)/ubi-utils/new-utils
 
 IGNORE=${wildcard $(BUILDDIR)/.*.c.dep}
 -include ${IGNORE}
@@ -56,8 +55,7 @@
 clean:
 	rm -f $(BUILDDIR)/*.o $(TARGETS) $(BUILDDIR)/.*.c.dep $(SYMLINKS)
 	if [ "$(BUILDDIR)x" != ".x" ]; then rm -rf $(BUILDDIR); fi
-	$(MAKE) -C $(BUILDDIR)/ubi-utils clean
-	$(MAKE) -C $(BUILDDIR)/mkfs.ubifs clean
+	$(MAKE) -C $(BUILDDIR)/ubi-utils/new-utils clean
 
 $(SYMLINKS):
 	ln -sf ../fs/jffs2/$@ $@
@@ -99,5 +97,4 @@
 	install -m0755 ${TARGETS} ${DESTDIR}/${SBINDIR}/
 	mkdir -p ${DESTDIR}/${MANDIR}/man1
 	gzip -9c mkfs.jffs2.1 > ${DESTDIR}/${MANDIR}/man1/mkfs.jffs2.1.gz
-	$(MAKE) -C $(BUILDDIR)/ubi-utils install
-	$(MAKE) -C $(BUILDDIR)/mkfs.ubifs install
+	$(MAKE) -C $(BUILDDIR)/ubi-utils/new-utils install
