--- mtd-utils.orig/ubi-utils/new-utils/src/ubinize.c	2008-09-23 21:10:28.000000000 +0100
+++ mtd-utils/ubi-utils/new-utils/src/ubinize.c	2008-09-23 20:54:37.000000000 +0100
@@ -561,20 +561,22 @@
 			goto out_free;
 		}
 
-		fd = open(img, O_RDONLY);
-		if (fd == -1) {
-			sys_errmsg("cannot open \"%s\"", img);
-			goto out_free;
-		}
-
-		verbose(args.verbose, "writing volume %d", vi[i].id);
-		verbose(args.verbose, "image file: %s", img);
-
-		err = ubigen_write_volume(&ui, &vi[i], args.ec, st.st_size, fd, args.out_fd);
-		close(fd);
-		if (err) {
-			errmsg("cannot write volume for section \"%s\"", sname);
-			goto out_free;
+		if (img) {
+			fd = open(img, O_RDONLY);
+			if (fd == -1) {
+				sys_errmsg("cannot open \"%s\"", img);
+				goto out_free;
+			}
+
+			verbose(args.verbose, "writing volume %d", vi[i].id);
+			verbose(args.verbose, "image file: %s", img);
+
+			err = ubigen_write_volume(&ui, &vi[i], args.ec, st.st_size, fd, args.out_fd);
+			close(fd);
+			if (err) {
+				errmsg("cannot write volume for section \"%s\"", sname);
+				goto out_free;
+			}
 		}
 
 		if (args.verbose)
