Index: linux-2.6.26/arch/arm/plat-mxc/usb_common.c
===================================================================
--- linux-2.6.26.orig/arch/arm/plat-mxc/usb_common.c	2009-04-29 16:14:36.000000000 +0100
+++ linux-2.6.26/arch/arm/plat-mxc/usb_common.c	2009-04-29 16:15:13.000000000 +0100
@@ -450,13 +450,14 @@
 		else
 			usbh1_set_serial_xcvr();
 	} else if (xops->xcvr_type == PORTSC_PTS_ULPI) {
+#ifdef CONFIG_USB_EHCI_ARC_H1
 		if (cpu_is_mx51()) {
 			usbh1_set_ulpi_xcvr();
 			gpio_usbh1_setback_stp();
 		} else
 			usbh2_set_ulpi_xcvr();
+#endif
 	}
-
 	pr_debug("%s: %s success\n", __func__, pdata->name);
 	return 0;
 }
