--- AR6kSDK.build_sw.62/host/wlan/include/ieee80211_node.h.orig	2010-05-17 13:34:21.000000000 +0200
+++ AR6kSDK.build_sw.62/host/wlan/include/ieee80211_node.h	2010-05-17 13:48:36.000000000 +0200
@@ -80,6 +80,7 @@ struct ieee80211_node_table {
 #endif
 };
 
-#define WLAN_NODE_INACT_TIMEOUT_MSEC            120000
+// 08/20/10 - fm - Bug: 13950 - Shorter scan list cache timeout.
+#define WLAN_NODE_INACT_TIMEOUT_MSEC            15000 // was 120000
 
 #endif /* _IEEE80211_NODE_H_ */
