--- AR6kSDK.build_sw.62/host/os/linux/include/config_linux.h.orig	2009-09-08 10:23:41.000000000 +0200
+++ AR6kSDK.build_sw.62/host/os/linux/include/config_linux.h	2009-09-08 10:24:06.000000000 +0200
@@ -37,7 +37,7 @@ extern "C" {
 /*
  * Host side Test Command support
  */
-#define CONFIG_HOST_TCMD_SUPPORT
+#undef CONFIG_HOST_TCMD_SUPPORT
 
 #define USE_4BYTE_REGISTER_ACCESS
 
--- AR6kSDK.build_sw.62/host/tools/tcmd/athtestcmd.c.orig	2009-09-08 10:24:55.000000000 +0200
+++ AR6kSDK.build_sw.62/host/tools/tcmd/athtestcmd.c	2009-09-08 10:25:41.000000000 +0200
@@ -121,6 +121,7 @@ usage(void)
     exit(-1);
 }
 
+#ifdef CONFIG_HOST_TCMD_SUPPORT
 int
 main (int argc, char **argv)
 {
@@ -343,6 +344,14 @@ main (int argc, char **argv)
 
     exit (0);
 }
+#else
+int
+main (int argc, char **argv)
+{
+    fprintf(stderr, "Testcommands disabled: Please compile with: CONFIG_HOST_TCMD_SUPPORT defined.\n");
+    exit(-1);
+}
+#endif
 
 static void
 rxReport(void *buf)
