--- AR6kSDK.build_sw.62/host/tools/eeprom/eeprom.c	2009-07-10 13:16:45.000000000 +0200
+++ AR6kSDK.build_sw.62/host/tools/eeprom/eeprom_fixed.c	2009-07-10 13:18:07.000000000 +0200
@@ -83,7 +83,7 @@ A_UINT32 target_version;
 A_UINT32 target_type;
 unsigned char BMI_read_reg_cmd[8];
 unsigned char BMI_write_reg_cmd[12];
-unsigned char BMI_read_mem_cmd[12];
+//unsigned char BMI_read_mem_cmd[12];
 unsigned char BMI_write_mem_cmd[BMI_DATASZ_MAX+3*sizeof(A_UINT32)];
 unsigned char mac_addr[ATH_MAC_LEN];
 unsigned short regcode;
@@ -238,6 +238,8 @@ BMI_write_reg(A_UINT32 address, A_UINT32
 void
 BMI_read_mem(A_UINT32 address, A_UINT32 *pvalue)
 {
+    unsigned char BMI_read_mem_cmd[12];
+
     DEBUG("BMI_read_mem address=0x%x\n", address);
 
     ((int *)BMI_read_mem_cmd)[0] = AR6000_XIOCTL_BMI_READ_MEMORY;
