Index: AR6kSDK.build_sw.62/host/hif/sdio/linux_sdio/include/hif_internal.h
===================================================================
--- AR6kSDK.build_sw.62.orig/host/hif/sdio/linux_sdio/include/hif_internal.h	2009-09-04 16:18:25.000000000 +0100
+++ AR6kSDK.build_sw.62/host/hif/sdio/linux_sdio/include/hif_internal.h	2009-09-04 18:18:49.000000000 +0100
@@ -58,7 +58,7 @@
 typedef struct bus_request {
     struct bus_request *next;       /* link list of available requests */
     struct bus_request *inusenext;  /* link list of in use requests */
-    struct semaphore sem_req;
+    struct compat_semaphore sem_req;
     A_UINT32 address;               /* request data */
     A_UCHAR *buffer;
     A_UINT32 length;
@@ -72,7 +72,7 @@
     void *htc_handle;
     spinlock_t asynclock;
     struct task_struct* async_task;             /* task to handle async commands */
-    struct semaphore sem_async;                 /* wake up for async task */
+    struct compat_semaphore sem_async;                 /* wake up for async task */
     int    async_shutdown;                      /* stop the async task */
     struct completion async_completion;          /* thread completion */
     BUS_REQUEST   *asyncreq;                    /* request for async tasklet */
