Index: SDL-1.2.13/src/video/fbcon/SDL_fbevents.c
===================================================================
--- SDL-1.2.13.orig/src/video/fbcon/SDL_fbevents.c	2012-06-04 10:06:04.681942493 -0700
+++ SDL-1.2.13/src/video/fbcon/SDL_fbevents.c	2012-06-04 10:07:05.712684790 -0700
@@ -356,6 +356,8 @@
 				if (tmp_fd > 0) {
 					sprintf(str, "echo -e '\033[9;0]' > %s", vtpath);
 					system(str);
+					sprintf(str, "echo 0 > /sys/class/graphics/fbcon/cursor_blink");
+					system(str);
 				}
 
 				/* This needs to be our controlling tty
