Index: SDL-1.2.13/src/video/fbcon/SDL_fbvideo.c
===================================================================
--- SDL-1.2.13.orig/src/video/fbcon/SDL_fbvideo.c	2009-04-21 09:58:46.000000000 +0100
+++ SDL-1.2.13/src/video/fbcon/SDL_fbvideo.c	2009-04-21 09:59:39.000000000 +0100
@@ -967,9 +967,9 @@
 		vinfo.green.length = vinfo.green.offset = 0;
 		vinfo.blue.length = vinfo.blue.offset = 0;
 		vinfo.transp.length = vinfo.transp.offset = 0;
-		if ( ! choose_fbmodes_mode(&vinfo) ) {
-			choose_vesa_mode(&vinfo);
-		}
+		//if ( ! choose_fbmodes_mode(&vinfo) ) {
+		//	choose_vesa_mode(&vinfo);
+		//}
 #ifdef FBCON_DEBUG
 		fprintf(stderr, "Printing wanted vinfo:\n");
 		print_vinfo(&vinfo);
