From 785f3ee2e1d1d9a09b26ea75f1b65f3388ea5b15 Mon Sep 17 00:00:00 2001
From: Josef 'Jeff' Sipek <jsipek@cs.sunysb.edu>
Date: Sat, 17 Mar 2007 19:42:25 -0400
Subject: [PATCH] MAINTAINERS: Unionfs URL update

Signed-off-by: Josef 'Jeff' Sipek <jsipek@cs.sunysb.edu>
---
 MAINTAINERS |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/MAINTAINERS b/MAINTAINERS
index 9172db8..42f93f7 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -4097,7 +4097,7 @@ UNIONFS
 P:	Josef "Jeff" Sipek
 M:	jsipek@cs.sunysb.edu
 L:	unionfs@filesystems.org
-W:	http://www.unionfs.org
+W:	http://unionfs.filesystems.org
 S:	Maintained
 
 USB ACM DRIVER
-- 
1.5.4.3

