From 1f9296069d3c1852ef423335642eed0707649d19 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Fri, 1 Jun 2007 00:47:20 -0400
Subject: [PATCH] remove unnecessary comment

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/file.c |    4 ----
 1 files changed, 0 insertions(+), 4 deletions(-)

diff --git a/fs/unionfs/file.c b/fs/unionfs/file.c
index 24a14b3..62da4e4 100644
--- a/fs/unionfs/file.c
+++ b/fs/unionfs/file.c
@@ -18,10 +18,6 @@
 
 #include "union.h"
 
-/*******************
- * File Operations *
- *******************/
-
 static ssize_t unionfs_read(struct file *file, char __user *buf,
 			    size_t count, loff_t *ppos)
 {
-- 
1.5.4.3

