From f6c45436c549a9e2d7f4faac17d666b944b362f1 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sun, 10 Jun 2007 17:51:42 -0400
Subject: [PATCH] minor indentation correction

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/lookup.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/lookup.c b/fs/unionfs/lookup.c
index 98a05a8..c65875d 100644
--- a/fs/unionfs/lookup.c
+++ b/fs/unionfs/lookup.c
@@ -490,7 +490,7 @@ int new_dentry_private_data(struct dentry *dentry)
 					    GFP_ATOMIC);
 	info = UNIONFS_D(dentry);
 	if (!info)
-	  goto out;
+		goto out;
 
 	mutex_init(&info->lock);
 	info->lower_paths = NULL;
-- 
1.5.4.3

