From 094552dcb4347a715055377277ae7fa7f8a80c0c Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sun, 19 Aug 2007 15:54:06 -0400
Subject: [PATCH] Unionfs: release 2.1.1.

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/Makefile b/fs/unionfs/Makefile
index e1cef40..f798b31 100644
--- a/fs/unionfs/Makefile
+++ b/fs/unionfs/Makefile
@@ -1,4 +1,4 @@
-UNIONFS_VERSION="2.1 (for 2.6.23-rc1)"
+UNIONFS_VERSION="2.1.1 (for 2.6.23-rc3)"
 
 EXTRA_CFLAGS += -DUNIONFS_VERSION=\"$(UNIONFS_VERSION)\"
 
-- 
1.5.4.3

