From 8eeeb2971f3f381462ebbbe990d6a4276b4dcc83 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Fri, 2 Nov 2007 19:33:00 -0400
Subject: [PATCH] Unionfs: revert fs/Kconfig text update

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/Kconfig |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/fs/Kconfig b/fs/Kconfig
index edbeb0b..ee32927 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -1052,7 +1052,8 @@ config UNION_FS
 	  Unionfs is a stackable unification file system, which appears to
 	  merge the contents of several directories (branches), while keeping
 	  their physical content separate.
-	  See <http://unionfs.filesystems.org/> for details.
+
+	  See <http://unionfs.filesystems.org> for details
 
 config UNION_FS_XATTR
 	bool "Unionfs extended attributes"
-- 
1.5.4.3

