From 15481ff0d58a76c4e3192d45d93c33d70f3c3674 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sun, 23 Dec 2007 16:36:29 -0500
Subject: [PATCH] Unionfs: initialize namelist variable in rename

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/rename.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/rename.c b/fs/unionfs/rename.c
index 1019d47..9306a2b 100644
--- a/fs/unionfs/rename.c
+++ b/fs/unionfs/rename.c
@@ -462,7 +462,7 @@ int unionfs_rename(struct inode *old_dir, struct dentry *old_dentry,
 		}
 
 		if (S_ISDIR(new_dentry->d_inode->i_mode)) {
-			struct unionfs_dir_state *namelist;
+			struct unionfs_dir_state *namelist = NULL;
 			/* check if this unionfs directory is empty or not */
 			err = check_empty(new_dentry, &namelist);
 			if (err)
-- 
1.5.4.3

