From ed943d5789ed5a598364795c596ae440d7ae1dac Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Thu, 18 Sep 2008 21:49:47 -0400
Subject: [PATCH] Unionfs: just warn in put_super instead of BUG-ing

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/super.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/super.c b/fs/unionfs/super.c
index 2f3c050..6fd97a8 100644
--- a/fs/unionfs/super.c
+++ b/fs/unionfs/super.c
@@ -124,7 +124,7 @@ static void unionfs_put_super(struct super_block *sb)
 			       bindex, branch_count(sb, bindex));
 			leaks = 1;
 		}
-	BUG_ON(leaks != 0);
+	WARN_ON(leaks != 0);
 
 	/* decrement lower super references */
 	for (bindex = bstart; bindex <= bend; bindex++) {
-- 
1.5.4.3

