From da5fd4f797ee54dcdc2e0ce62b88a1a53db7a1fd Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 28 Apr 2007 21:04:05 -0400
Subject: [PATCH] mmap: remove defunct comments

---
 fs/unionfs/mmap.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/mmap.c b/fs/unionfs/mmap.c
index 7d27bd8..3de7757 100644
--- a/fs/unionfs/mmap.c
+++ b/fs/unionfs/mmap.c
@@ -201,7 +201,7 @@ int unionfs_commit_write(struct file *file, struct page *page, unsigned from,
 	if ((err = unionfs_file_revalidate(file, 1)))
 		goto out;
 
-	inode = page->mapping->host;	/* CPW: Moved below print_entry_location */
+	inode = page->mapping->host;
 	lower_inode = unionfs_lower_inode(inode);
 
 	if (UNIONFS_F(file) != NULL)
@@ -272,7 +272,7 @@ void unionfs_sync_page(struct page *page)
 	struct page *lower_page;
 	struct address_space *mapping = page->mapping;
 
-	inode = page->mapping->host;	/* CPW: Moved below print_entry_location */
+	inode = page->mapping->host;
 	lower_inode = unionfs_lower_inode(inode);
 
 	/* find lower page (returns a locked page) */
-- 
1.5.4.3

