From 304b095e796e9cb6fd08eb673505f68059714cbd Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Thu, 18 Sep 2008 21:49:47 -0400
Subject: [PATCH] Unionfs: fix off-by-one but in mkdir

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/inode.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 2ad8c54..84efb30 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -517,7 +517,7 @@ static int unionfs_mkdir(struct inode *parent, struct dentry *dentry, int mode)
 		if (err)
 			break;
 
-		for (i = bindex + 1; i < bend; i++) {
+		for (i = bindex + 1; i <= bend; i++) {
 			if (unionfs_lower_dentry_idx(dentry, i)) {
 				dput(unionfs_lower_dentry_idx(dentry, i));
 				unionfs_set_lower_dentry_idx(dentry, i, NULL);
-- 
1.5.4.3

