From 940da53dc5ac75a96ae99bd67c4f085842e0c7d2 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 19 May 2007 20:44:25 -0400
Subject: [PATCH] d_splice bugfix: update dentry in our lookup if needed.

---
 fs/unionfs/inode.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index d3e0a4f..fcdd739 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -266,8 +266,11 @@ static struct dentry *unionfs_lookup(struct inode *parent,
 		nd->dentry = path_save.dentry;
 		nd->mnt = path_save.mnt;
 	}
-	if (!IS_ERR(ret))
+	if (!IS_ERR(ret)) {
+		if (ret)
+			dentry = ret;
 		unionfs_inherit_mnt(dentry);
+	}
 
 	unionfs_check_inode(parent);
 	unionfs_check_dentry(dentry);
-- 
1.5.4.3

