From 175fa27aa4e392113e8e4514ee0218200f7fc4d8 Mon Sep 17 00:00:00 2001
From: Josef 'Jeff' Sipek <jsipek@cs.sunysb.edu>
Date: Mon, 3 Sep 2007 17:32:07 -0400
Subject: [PATCH] Unionfs: Remove unused #defines

Signed-off-by: Josef 'Jeff' Sipek <jsipek@cs.sunysb.edu>
---
 fs/unionfs/union.h |    4 ----
 1 files changed, 0 insertions(+), 4 deletions(-)

diff --git a/fs/unionfs/union.h b/fs/unionfs/union.h
index f03143e..53327f2 100644
--- a/fs/unionfs/union.h
+++ b/fs/unionfs/union.h
@@ -441,10 +441,6 @@ static inline int is_robranch(const struct dentry *dentry)
 #define UNIONFS_DIR_OPAQUE_NAME "__dir_opaque"
 #define UNIONFS_DIR_OPAQUE UNIONFS_WHPFX UNIONFS_DIR_OPAQUE_NAME
 
-#ifndef DEFAULT_POLLMASK
-#define DEFAULT_POLLMASK (POLLIN | POLLOUT | POLLRDNORM | POLLWRNORM)
-#endif /* not DEFAULT_POLLMASK */
-
 /*
  * EXTERNALS:
  */
-- 
1.5.4.3

