From 29c8dc8a9fe6f7e147f7256a3474f3805d33704b Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Thu, 27 Dec 2007 19:04:09 -0500
Subject: [PATCH] Unionfs: cleanup lower inodes after successful unlink

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/unlink.c |   11 +++++++++--
 1 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/unlink.c b/fs/unionfs/unlink.c
index 1e81494..a1c82b6 100644
--- a/fs/unionfs/unlink.c
+++ b/fs/unionfs/unlink.c
@@ -91,7 +91,9 @@ out:
 int unionfs_unlink(struct inode *dir, struct dentry *dentry)
 {
 	int err = 0;
+	struct inode *inode = dentry->d_inode;
 
+	BUG_ON(S_ISDIR(inode->i_mode));
 	unionfs_read_lock(dentry->d_sb, UNIONFS_SMUTEX_CHILD);
 	unionfs_lock_dentry(dentry, UNIONFS_DMUTEX_CHILD);
 
@@ -104,8 +106,13 @@ int unionfs_unlink(struct inode *dir, struct dentry *dentry)
 	err = unionfs_unlink_whiteout(dir, dentry);
 	/* call d_drop so the system "forgets" about us */
 	if (!err) {
-		if (!S_ISDIR(dentry->d_inode->i_mode))
-			unionfs_postcopyup_release(dentry);
+		unionfs_postcopyup_release(dentry);
+		if (inode->i_nlink == 0) {
+			/* drop lower inodes */
+			iput(unionfs_lower_inode(inode));
+			unionfs_set_lower_inode(inode, NULL);
+			ibstart(inode) = ibend(inode) = -1;
+		}
 		d_drop(dentry);
 		/*
 		 * if unlink/whiteout succeeded, parent dir mtime has
-- 
1.5.4.3

