From 88a51e243bfe639ca4d5213cdf235b4091c08f4a Mon Sep 17 00:00:00 2001
From: Yiannis Pericleous <yiannos@louie.fsl.cs.sunysb.edu>
Date: Sat, 28 Apr 2007 22:04:54 -0400
Subject: [PATCH] mmap: coding style: reformated a line longer than 80 chars

---
 fs/unionfs/mmap.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/mmap.c b/fs/unionfs/mmap.c
index 3de7757..99b3b2f 100644
--- a/fs/unionfs/mmap.c
+++ b/fs/unionfs/mmap.c
@@ -220,7 +220,8 @@ int unionfs_commit_write(struct file *file, struct page *page, unsigned from,
 	 */
 	old_fs = get_fs();
 	set_fs(KERNEL_DS);
-	err = vfs_write(lower_file, page_data + from, bytes, &lower_file->f_pos);
+	err = vfs_write(lower_file, page_data + from, bytes,
+			&lower_file->f_pos);
 	set_fs(old_fs);
 
 	kunmap(page);
-- 
1.5.4.3

