From 2ca4933202c8e791c94be7ea7458df360b4e2919 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Thu, 18 Sep 2008 21:49:47 -0400
Subject: [PATCH] Unionfs: drop branchcount on error

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/commonfops.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/fs/unionfs/commonfops.c b/fs/unionfs/commonfops.c
index 5861970..51f10a2 100644
--- a/fs/unionfs/commonfops.c
+++ b/fs/unionfs/commonfops.c
@@ -181,6 +181,7 @@ static int open_all_files(struct file *file)
 				    unionfs_lower_mnt_idx(dentry, bindex),
 				    file->f_flags);
 		if (IS_ERR(lower_file)) {
+			branchput(sb, bindex);
 			err = PTR_ERR(lower_file);
 			goto out;
 		} else {
-- 
1.5.4.3

