Index: linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c
===================================================================
--- linux-2.6.26.orig/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-07-24 16:49:06.000000000 -0700
+++ linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-07-24 16:53:41.000000000 -0700
@@ -647,6 +647,15 @@
 	mxc_request_iomux(MX25_PIN_LSCLK, MUX_CONFIG_FUNC);
 	mxc_request_iomux(MX25_PIN_OE_ACD, MUX_CONFIG_FUNC);
 
+	/* LCD Mode control */
+	mxc_request_iomux(MX25_PIN_CSPI1_MOSI, MUX_CONFIG_GPIO);
+	mxc_request_iomux(MX25_PIN_CSPI1_MISO, MUX_CONFIG_GPIO);
+	mxc_set_gpio_direction(MX25_PIN_CSPI1_MISO, 0);
+	mxc_set_gpio_direction(MX25_PIN_CSPI1_MOSI, 0);
+	mxc_set_gpio_dataout(MX25_PIN_CSPI1_MISO, 1);
+	mxc_set_gpio_dataout(MX25_PIN_CSPI1_MOSI, 0);
+
+
 	/* LCD Brightness controls */
 	mxc_request_iomux(MX25_PIN_CONTRAST, MUX_CONFIG_FUNC); /* Brightness */
 	mxc_request_iomux(MX25_PIN_D11, MUX_CONFIG_GPIO);      /* extra low brightness mode.  0 = bright, 1 = dim */
@@ -660,7 +669,7 @@
 		mxc_request_iomux(BABY_SLCDC_CS,  MUX_CONFIG_GPIO);
 	}
 
-#define LCD_PAD_CTL (PAD_CTL_PKE_ENABLE | PAD_CTL_PUE_PUD | PAD_CTL_100K_PU)
+#define LCD_PAD_CTL (PAD_CTL_PKE_ENABLE | PAD_CTL_PUE_PUD | PAD_CTL_100K_PU | PAD_CTL_SRE_FAST | PAD_CTL_DRV_MAX)
 	mxc_iomux_set_pad(MX25_PIN_LD0, LCD_PAD_CTL);
 	mxc_iomux_set_pad(MX25_PIN_LD1, LCD_PAD_CTL);
 	mxc_iomux_set_pad(MX25_PIN_LD2, LCD_PAD_CTL);
@@ -681,7 +690,7 @@
 	mxc_iomux_set_pad(MX25_PIN_GPIO_F, LCD_PAD_CTL);
 	mxc_iomux_set_pad(MX25_PIN_HSYNC, LCD_PAD_CTL);
 	mxc_iomux_set_pad(MX25_PIN_VSYNC, LCD_PAD_CTL);
-	mxc_iomux_set_pad(MX25_PIN_LSCLK, LCD_PAD_CTL | PAD_CTL_SRE_FAST);
+	mxc_iomux_set_pad(MX25_PIN_LSCLK, PAD_CTL_SRE_SLOW | PAD_CTL_DRV_NORMAL);
 	mxc_iomux_set_pad(MX25_PIN_OE_ACD, LCD_PAD_CTL);
 	mxc_iomux_set_pad(MX25_PIN_CONTRAST, LCD_PAD_CTL);
 	mxc_iomux_set_pad(MX25_PIN_D11,  PAD_CTL_DRV_3_3V | PAD_CTL_PKE_ENABLE | PAD_CTL_PUE_KEEPER);
@@ -728,6 +737,8 @@
 	mxc_request_gpio(MX25_PIN_LSCLK);
 	mxc_request_gpio(MX25_PIN_OE_ACD);
 
+	mxc_free_iomux(MX25_PIN_CSPI1_MOSI, MUX_CONFIG_GPIO);
+	mxc_free_iomux(MX25_PIN_CSPI1_MISO, MUX_CONFIG_GPIO);
 	mxc_free_iomux(MX25_PIN_LD0, MUX_CONFIG_GPIO);
 	mxc_free_iomux(MX25_PIN_LD1, MUX_CONFIG_GPIO);
 	mxc_free_iomux(MX25_PIN_LD2, MUX_CONFIG_GPIO);
