Index: linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c
===================================================================
--- linux-2.6.26.orig/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-09-14 16:40:45.000000000 +0100
+++ linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-09-14 16:38:58.000000000 +0100
@@ -38,6 +38,7 @@
 {
 	/* MCLK uses CLKO */
 	mxc_request_iomux(MX25_PIN_CLKO,  	    MUX_CONFIG_FUNC);
+	mxc_iomux_set_pad(MX25_PIN_CLKO, PAD_CTL_DRV_NORMAL);
 
 	mxc_request_iomux(BABY_DAC_SDIN,  	    MUX_CONFIG_ALT1);
 	mxc_request_iomux(BABY_DAC_SDOUT, 	    MUX_CONFIG_ALT1);
