Index: linux-2.6.26/drivers/mxc/baby/msp430/msp430.h
===================================================================
--- linux-2.6.26.orig/drivers/mxc/baby/msp430/msp430.h	2012-11-05 08:25:50.777137955 -0800
+++ linux-2.6.26/drivers/mxc/baby/msp430/msp430.h	2012-10-23 16:46:19.139330000 -0700
@@ -68,6 +68,8 @@
 #define I2C_TIME_REGISTER                70
 #define I2C_ALARM_REGISTER               71
 #define I2C_KILL_REGISTER                99
+#define I2C_MEMORY_ADDRESS 		126
+#define I2C_MEMORY_DATA			127
 
 //  Test registers.
 //  Result values should be:
Index: linux-2.6.26/drivers/mxc/baby/msp430/msp430_i2c.c
===================================================================
--- linux-2.6.26.orig/drivers/mxc/baby/msp430/msp430_i2c.c	2012-11-05 08:25:51.297139664 -0800
+++ linux-2.6.26/drivers/mxc/baby/msp430/msp430_i2c.c	2012-10-23 16:55:54.503473000 -0700
@@ -242,6 +242,8 @@
 static SENSOR_DEVICE_ATTR_2(adc_13		 ,   S_IRUGO, msp430_i2c_show, NULL, 2, I2C_ADC_BASE+13);
 static SENSOR_DEVICE_ATTR_2(adc_14		 ,   S_IRUGO, msp430_i2c_show, NULL, 2, I2C_ADC_BASE+14);
 static SENSOR_DEVICE_ATTR_2(adc_15		 ,   S_IRUGO, msp430_i2c_show, NULL, 2, I2C_ADC_BASE+15);
+static SENSOR_DEVICE_ATTR_2(memory_address	 ,   S_IRUGO | S_IWUSR, msp430_i2c_show, msp430_i2c_store, 2, I2C_MEMORY_ADDRESS);
+static SENSOR_DEVICE_ATTR_2(memory_data 	 ,   S_IRUGO | S_IWUSR, msp430_i2c_show, msp430_i2c_store, 2, I2C_MEMORY_DATA);
 
 
 
@@ -289,6 +291,8 @@
 		&sensor_dev_attr_adc_13.dev_attr.attr,
 		&sensor_dev_attr_adc_14.dev_attr.attr,
 		&sensor_dev_attr_adc_15.dev_attr.attr,
+		&sensor_dev_attr_memory_address.dev_attr.attr,
+		&sensor_dev_attr_memory_data.dev_attr.attr,
                 NULL,
         },
 };
