Index: busybox_1_7_stable/networking/ifupdown.c
===================================================================
--- busybox_1_7_stable/networking/ifupdown.c	(revision 835)
+++ busybox_1_7_stable/networking/ifupdown.c	(revision 836)
@@ -526,7 +526,8 @@
 		return 0;
 #endif
-	return execute("udhcpc -R -n -p /var/run/udhcpc.%iface%.pid "
+	return execute("udhcpc -R -p /var/run/udhcpc.%iface%.pid "
+					"-b --syslog "
			"-i %iface%[[ -H %hostname%]][[ -c %clientid%]][[ -s %script%]][[ %udhcpc_opts%]]",
			ifd, exec);
 }
 #else

