--- alac_decoder/main.c.orig	2006-09-29 16:15:23.000000000 -0700
+++ alac_decoder/main.c	2009-03-02 19:20:18.000000000 -0800
@@ -38,6 +38,9 @@
 #include <stdlib.h>
 #include <errno.h>
 #include <string.h>
+#ifdef WIN32
+#include <fcntl.h>
+#endif
 
 #include "demux.h"
 #include "decomp.h"
@@ -145,6 +148,10 @@
             fprintf(stderr, "read %i bytes. total: %i\n", outputBytes, bytes_read);
 
         fwrite(pDestBuffer, outputBytes, 1, output_file);
+        if (ferror(output_file)) {
+	          perror("GetBuffer fwrite error");
+        	break;
+    }
     }
     if (verbose)
         fprintf(stderr, "done reading, read %i frames\n", i);
@@ -173,7 +180,7 @@
                     "                    other m4a file types.\n"
                     "\n"
                     "This software is Copyright (c) 2005 David Hammerton\n"
-                    "All rights reserved\n"
+                    "All rights reserved -  V 0.patched for Slimserver\n"
                     "http://crazney.net/\n");
     exit(1);
 }
