From 4aee46bbd0f2cf6873fa81c2190c91aa65c4a3c0 Mon Sep 17 00:00:00 2001
From: Artem Bityutskiy <Artem.Bityutskiy@nokia.com>
Date: Mon, 13 Oct 2008 10:25:41 +0300
Subject: [PATCH] UBIFS: slight compat code amendment

Signed-off-by: Artem Bityutskiy <Artem.Bityutskiy@nokia.com>
---
 fs/ubifs/compat.c |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/fs/ubifs/compat.c b/fs/ubifs/compat.c
index 8c61da0..4c32bdc 100644
--- a/fs/ubifs/compat.c
+++ b/fs/ubifs/compat.c
@@ -107,8 +107,8 @@ void ubifs_read_inode(struct inode *inode)
 	inode->i_mtime.tv_nsec = le32_to_cpu(ino->mtime_nsec);
 	inode->i_ctime.tv_sec  = (int64_t)le64_to_cpu(ino->ctime_sec);
 	inode->i_ctime.tv_nsec = le32_to_cpu(ino->ctime_nsec);
-	inode->i_mode  = le32_to_cpu(ino->mode);
-	inode->i_size  = le64_to_cpu(ino->size);
+	inode->i_mode = le32_to_cpu(ino->mode);
+	inode->i_size = le64_to_cpu(ino->size);
 
 	ui->data_len    = le32_to_cpu(ino->data_len);
 	ui->flags       = le32_to_cpu(ino->flags);
@@ -118,6 +118,7 @@ void ubifs_read_inode(struct inode *inode)
 	ui->xattr_size  = le64_to_cpu(ino->xattr_size);
 	ui->xattr_names = le32_to_cpu(ino->xattr_names);
 	ui->synced_i_size = ui->ui_size = inode->i_size;
+
 	ui->xattr = (ui->flags & UBIFS_XATTR_FL) ? 1 : 0;
 
 	err = validate_inode(c, inode);
@@ -156,7 +157,7 @@ void ubifs_read_inode(struct inode *inode)
 			err = 12;
 			goto out_invalid;
 		}
-		ui->data = kmalloc(ui->data_len + 1, GFP_KERNEL);
+		ui->data = kmalloc(ui->data_len + 1, GFP_NOFS);
 		if (!ui->data) {
 			err = -ENOMEM;
 			goto out_ino;
-- 
1.5.6.3

