From 00219fb6218d0395ffc8a3ed57d1c974472b7926 Mon Sep 17 00:00:00 2001
From: Artem Bityutskiy <Artem.Bityutskiy@nokia.com>
Date: Wed, 3 Dec 2008 14:55:57 +0200
Subject: [PATCH] UBIFS: remove more compat left-overs

Signed-off-by: Artem Bityutskiy <Artem.Bityutskiy@nokia.com>
---
 fs/ubifs/compat.h |    6 ------
 1 files changed, 0 insertions(+), 6 deletions(-)

diff --git a/fs/ubifs/compat.h b/fs/ubifs/compat.h
index 6d42bcf..2eb28ab 100644
--- a/fs/ubifs/compat.h
+++ b/fs/ubifs/compat.h
@@ -25,12 +25,6 @@
 
 #include <linux/version.h>
 
-struct inode;
-struct file;
-struct page;
-struct ubifs_info;
-struct retries_info;
-
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24))
 #define bdi_init(x) 0
 #define bdi_destroy(x)
-- 
1.5.6.3

