Index: alsa-utils-1.0.18/Makefile.am
===================================================================
--- alsa-utils-1.0.18.orig/Makefile.am	2008-10-29 12:48:00.000000000 +0000
+++ alsa-utils-1.0.18/Makefile.am	2008-11-26 12:45:55.000000000 +0000
@@ -9,8 +9,8 @@
 else
 ALSACONF_DIR=
 endif
-SUBDIRS= include alsactl $(ALSACONF_DIR) $(ALSAMIXER_DIR) amidi amixer \
-	 aplay iecset seq speaker-test utils m4 po
+SUBDIRS= include alsactl $(ALSACONF_DIR) $(ALSAMIXER_DIR) amixer \
+	 aplay iecset speaker-test utils m4 po
 EXTRA_DIST= config.rpath  TODO gitcompile
 AUTOMAKE_OPTIONS=foreign
 ACLOCAL_AMFLAGS = -I m4
