--- src/main.c.orig	2009-05-26 14:46:20.000000000 +0200
+++ src/main.c	2009-05-26 14:44:12.000000000 +0200
@@ -511,6 +511,15 @@
   frame.udp.source  = htons(DHCP_PORT_SERVER);
   frame.udp.dest    = htons(DHCP_PORT_CLIENT);
 
+  {
+    char temp[255];
+    struct in_addr ip;
+
+    ip.s_addr = header->yiaddr;
+    sprintf( temp, "/usr/sbin/bridge_helper.sh %s", inet_ntoa( ip));
+    system( temp);
+  }
+
   sendEtherFrame(iface, &frame, buffer, size);
 }
 
