--- busybox-1.18.2/networking/udhcp/dhcpc.c.orig	2011-01-16 19:21:31.000000000 +0100
+++ busybox-1.18.2/networking/udhcp/dhcpc.c	2011-01-31 21:14:24.000000000 +0100
@@ -1118,12 +1118,14 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 	bb_sanitize_stdio();
 	/* Equivalent of doing a fflush after every \n */
 	setlinebuf(stdout);
-	/* Create pidfile */
-	write_pidfile(client_config.pidfile);
 	/* Goes to stdout (unless NOMMU) and possibly syslog */
 	bb_info_msg("%s (v"BB_VER") started", applet_name);
 	/* Set up the signal pipe */
 	udhcp_sp_setup();
+
+	/* Create pidfile */
+	write_pidfile(client_config.pidfile);
+
 	/* We want random_xid to be random... */
 	srand(monotonic_us());
 
