Index: busybox-1.13.3/networking/udhcp/dhcpc.c
===================================================================
--- busybox-1.13.3.orig/networking/udhcp/dhcpc.c	2009-04-08 12:35:02.000000000 +0100
+++ busybox-1.13.3/networking/udhcp/dhcpc.c	2009-04-08 12:35:08.000000000 +0100
@@ -662,6 +662,14 @@
 				timeout = 0;
 				packet_num = 0;
 				already_waited_sec = 0;
+
+/* The Microsoft MN-700 router appears to have a problem with hostnames
+ * larger than 16 characters, and responds with a DHCP NAK. To workaround
+ * this, on a NAK we trucate the hostname is applicable.
+ */
+				if (client_config.hostname && client_config.hostname[OPT_LEN] > 16) {
+					client_config.hostname[OPT_LEN] = 16;
+				}
 			}
 			continue;
 		/* case BOUND, RELEASED: - ignore all packets */
