--- busybox-1.18.2/networking/ifupdown.c.orig	2011-01-16 19:19:54.000000000 +0100
+++ busybox-1.18.2/networking/ifupdown.c	2011-01-31 20:57:47.000000000 +0100
@@ -546,7 +546,8 @@ static int FAST_FUNC dhcp_up(struct inte
 	if (!execute("ifconfig %iface%[[ hw %hwaddress%]] up", ifd, exec))
 		return 0;
 #  endif
-	return execute("udhcpc " UDHCPC_CMD_OPTIONS " -p /var/run/udhcpc.%iface%.pid "
+	return execute("udhcpc -R -a -p /var/run/udhcpc.%iface%.pid "
+					"-b --syslog "
 			"-i %iface%[[ -H %hostname%]][[ -c %client%]][[ -s %script%]][[ %udhcpc_opts%]]",
 			ifd, exec);
 }

