Index: linux-2.6.26/drivers/char/hw_random/fsl-rngc.c
===================================================================
--- linux-2.6.26.orig/drivers/char/hw_random/fsl-rngc.c	2009-04-28 11:15:50.000000000 +0100
+++ linux-2.6.26/drivers/char/hw_random/fsl-rngc.c	2009-04-28 12:13:56.000000000 +0100
@@ -235,7 +235,7 @@
 		__raw_writel(cmd | RNGC_CMD_SELF_TEST,
 			rngc_base + RNGC_COMMAND);
 
-		wait_for_completion(&rng_self_testing);
+		wait_for_completion_timeout(&rng_self_testing, 100);
 
 	} while (__raw_readl(rngc_base + RNGC_ERROR) &
 		RNGC_ERROR_STATUS_ST_ERR);
@@ -254,7 +254,7 @@
 		cmd = __raw_readl(rngc_base + RNGC_COMMAND);
 		__raw_writel(cmd | RNGC_CMD_SEED, rngc_base + RNGC_COMMAND);
 
-		wait_for_completion(&rng_seed_done);
+		wait_for_completion_timeout(&rng_seed_done, 100);
 
 	} while (__raw_readl(rngc_base + RNGC_ERROR) &
 		RNGC_ERROR_STATUS_STAT_ERR);
