From 0ba5c427c89883ed6d719ba3a30137f31b359c5d Mon Sep 17 00:00:00 2001
From: Rob Herring <r.herring@freescale.com>
Date: Mon, 8 Sep 2008 13:35:19 -0500
Subject: [PATCH] ENGR00069679-2 pata_fsl: add mx51 kconfig dependency

---
 drivers/ata/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/ata/Kconfig b/drivers/ata/Kconfig
index 4a62608..bc43fd3 100644
--- a/drivers/ata/Kconfig
+++ b/drivers/ata/Kconfig
@@ -725,7 +725,7 @@ config PATA_BF54X
 
 config PATA_FSL
        tristate "Freescale on-chip PATA support"
-       depends on (ARCH_MX37 || ARCH_MX35 || ARCH_MX3 || ARCH_MX27)
+       depends on (ARCH_MX51 || ARCH_MX37 || ARCH_MX35 || ARCH_MX3 || ARCH_MX27)
        help
          On Freescale processors, say Y here if you wish to use the on-chip
          ATA interface.
-- 
1.5.4.4

