From f543680387bc3a13296fa710e0e4cdcaf81333da Mon Sep 17 00:00:00 2001
From: Artem Bityutskiy <Artem.Bityutskiy@nokia.com>
Date: Sat, 25 Oct 2008 18:54:04 +0300
Subject: [PATCH] MAINTAINERS: change UBI/UBIFS git tree URLs

Signed-off-by: Artem Bityutskiy <Artem.Bityutskiy@nokia.com>
---
 MAINTAINERS |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/MAINTAINERS b/MAINTAINERS
index 3882f9c..5d7b1a6 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -2761,7 +2761,7 @@ P:	Artem Bityutskiy
 M:	dedekind@infradead.org
 W:	http://www.linux-mtd.infradead.org/
 L:	linux-mtd@lists.infradead.org
-T:	git git://git.infradead.org/~dedekind/ubi-2.6.git
+T:	git git://git.infradead.org/ubi-2.6.git
 S:	Maintained
 
 MICROTEK X6 SCANNER
-- 
1.5.4.3

