Index: files/watchdog.conf
===================================================================
--- files.orig/watchdog.conf	2010-08-05 19:52:10.890096455 -0700
+++ files/watchdog.conf	2010-08-05 19:52:20.094079526 -0700
@@ -1,6 +1,7 @@
 
 # Check squeezeplay and squeezecenter
 pidfile			= /var/run/squeezeplay.pid
+pidfile                 = /var/run/monitor_msp430.pid
 
 # Note that this is the number of pages!
 min-memory		= 0
