diff -Naur linux-2.6.26.clean/drivers/mmc/host/mx_sdhci.c linux-2.6.26.caleb.cleaned/drivers/mmc/host/mx_sdhci.c
--- linux-2.6.26.clean/drivers/mmc/host/mx_sdhci.c	2009-03-19 03:23:46.000000000 -0700
+++ linux-2.6.26.caleb.cleaned/drivers/mmc/host/mx_sdhci.c	2009-03-14 22:21:22.000000000 -0700
@@ -1698,7 +1712,7 @@
 	host->detect_irq = platform_get_irq(pdev, 1);
 	if (!host->detect_irq) {
 		//host->flags &= ~SDHCI_CD_PRESENT;
-		host->flags |= ~SDHCI_CD_PRESENT;
+		host->flags |= SDHCI_CD_PRESENT;
 		if ((pdev->id >= 0) && (pdev->id < MXC_SDHCI_NUM))
 			mxc_fix_chips[pdev->id] = chip;
 		goto no_detect_irq;
