Index: linux-2.6.26/sound/soc/codecs/tlv320aic3104.c
===================================================================
--- linux-2.6.26.orig/sound/soc/codecs/tlv320aic3104.c	2009-05-08 20:44:49.000000000 +0100
+++ linux-2.6.26/sound/soc/codecs/tlv320aic3104.c	2009-05-08 20:45:00.000000000 +0100
@@ -189,10 +189,33 @@
 	/* DAC signal path */
 	7, ((1 << 7) |            /* fs(ref) = 44.1kHz */
 	    (1 << 3) |            /* Left-DAC plays left-channel data */
-	    (1 << 1)              /* Right-DAC plays right-channel data */
-	    ),
-	37, ((1 << 7) |           /* Left DAC is powered up */
-	     (1 << 6)             /* Right DAC is powered up */
+	    (1 << 1)),            /* Right-DAC plays right-channel data */
+	14, (1 << 7),             /* headphones = ac-coupled */
+	15, ((1    << 7) |        /* Left ADC Not muted */
+	     (0x28 << 0) |        /* Left ADC Gain = 20 dB */
+	     0),
+	16, ((1    << 7) |        /* Right ADC Not muted */
+	     (0x28 << 0) |        /* Right ADC Gain = 20 dB */
+	     0),
+	19, ((0 << 7) |           /* LINEIN1LP SE mode */
+	     (0 << 3) |           /* 0 dB connection to left ADC PGA Mixer */
+	     (1 << 2) |           /* Left ADC channel powered up */
+	     0),
+	22, ((0 << 7) |           /* LINEIN1RP SE mode */
+	     (0 << 3) |           /* 0 dB connection to right ADC PGA Mixer */
+	     (1 << 2) |           /* Right ADC channel powered up */
+	     0),
+	46, ((1 << 7) |           /* PGA_L routed to HPLOUT   */
+	     (0 << 0)),           /* With a gain of 0 dB      */
+	63, ((1 << 7) |           /* PGA_R routed to HPROUT   */
+	     (0 << 0)),           /* With a gain of 0 dB      */
+	81, ((1 << 7) |           /* PGA_L routed to LEFT_LOP */
+	     (0 << 0)),           /* With a gain of 0dB       */
+	91, ((1 << 7) |           /* PGA_R routed to RIGHT_LOP */
+	     (0 << 0)),           /* With a gain of 0dB       */
+
+	37, ((1 << 7) |           /* Left DAC is powered up   */
+	     (1 << 6)             /* Right DAC is powered up  */
 	     ),
 
 #if 0
