Index: linux-2.6.26/drivers/net/fec.c
===================================================================
--- linux-2.6.26.orig/drivers/net/fec.c	2009-07-02 12:48:55.000000000 +0100
+++ linux-2.6.26/drivers/net/fec.c	2009-07-02 12:48:56.000000000 +0100
@@ -1413,6 +1413,7 @@
 		/* test register: set TX clock to an output, set RMII mode */
 		{ mk_mii_write(25, ((0<<11)|(1<<10))), NULL },
 		{ mk_mii_write(MII_REG_CR, 0x1200), NULL }, /* autonegotiate */
+		{ mk_mii_write(MII_REG_ANAR, (1<<6) | (1<<5) | (0x01 << 0)), NULL }, /* autonegotiate 10BT only */
 		{ mk_mii_read(MII_REG_SR), mii_parse_sr },
 		{ mk_mii_end, }
 	},
