Index: linux-2.6.26/drivers/mmc/host/mx_sdhci.c
===================================================================
--- linux-2.6.26.orig/drivers/mmc/host/mx_sdhci.c	2009-03-20 12:01:28.000000000 +0000
+++ linux-2.6.26/drivers/mmc/host/mx_sdhci.c	2009-03-20 12:02:14.000000000 +0000
@@ -1696,7 +1697,7 @@
 	}
 	host->detect_irq = platform_get_irq(pdev, 1);
 	if (!host->detect_irq) {
-		host->flags &= ~SDHCI_CD_PRESENT;
+		host->flags |= SDHCI_CD_PRESENT;
 		if ((pdev->id >= 0) && (pdev->id < MXC_SDHCI_NUM))
 			mxc_fix_chips[pdev->id] = chip;
 		goto no_detect_irq;
