Index: linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c
===================================================================
--- linux-2.6.26.orig/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-05-31 05:21:19.000000000 +0800
+++ linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-05-31 05:21:39.000000000 +0800
@@ -97,7 +97,7 @@
 
 
 	/* MSP430 pins */
-	if (1) {
+	if (system_rev < 2 ) {
 		mxc_request_iomux   (BABY_MSP430_CHARGE_PUMP_PB1 , MUX_CONFIG_GPIO);
 	}
 	else {
