From 8032d3164f934a01464f902393c8b2eb234c40b5 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 19 May 2007 16:40:16 -0400
Subject: [PATCH] cleanup: consistent printk messages

---
 fs/unionfs/commonfops.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/commonfops.c b/fs/unionfs/commonfops.c
index 2b2aff8..fd2e546 100644
--- a/fs/unionfs/commonfops.c
+++ b/fs/unionfs/commonfops.c
@@ -399,8 +399,8 @@ int unionfs_file_revalidate(struct file *file, int willwrite)
 	if (willwrite && IS_WRITE_FLAG(file->f_flags) &&
 	    !IS_WRITE_FLAG(unionfs_lower_file(file)->f_flags) &&
 	    is_robranch(dentry)) {
-		printk(KERN_DEBUG "unionfs: Doing delayed copyup of a "
-		       "read-write file on a read-only branch.\n");
+		printk(KERN_DEBUG "unionfs: doing delayed copyup of a "
+		       "read-write file on a read-only branch\n");
 		err = do_delayed_copyup(file);
 	}
 
-- 
1.5.4.3

