From 413197dd73a38d6cf4f3612c5d6262662aa30784 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sat, 19 May 2007 17:14:27 -0400
Subject: [PATCH] debug: remove unnecessarily verbose debugging printk's

---
 fs/unionfs/debug.c |    5 -----
 1 files changed, 0 insertions(+), 5 deletions(-)

diff --git a/fs/unionfs/debug.c b/fs/unionfs/debug.c
index 161723e..ba33b48 100644
--- a/fs/unionfs/debug.c
+++ b/fs/unionfs/debug.c
@@ -43,8 +43,6 @@ void __unionfs_check_inode(const struct inode *inode,
   struct super_block *sb;
   int printed_caller = 0;
 
-  printk("_PCi(%ld:%o):%p:%s:%s:%d\n",
-	 inode->i_ino,inode->i_mode,inode,fname,fxn,line);
   /* for inodes now */
   BUG_ON(!inode);
   sb = inode->i_sb;
@@ -114,8 +112,6 @@ void __unionfs_check_dentry(const struct dentry *dentry,
   struct vfsmount *lower_mnt;
   int printed_caller = 0;
 
-  printk("_PCd(%s):%p:%p:%s:%s:%d\n",dentry->d_name.name,dentry,dentry->d_inode,
-	 fname,fxn,line);
   BUG_ON(!dentry);
   sb = dentry->d_sb;
   inode = dentry->d_inode;
@@ -274,7 +270,6 @@ void __unionfs_check_file(const struct file *file,
   struct super_block *sb;
   int printed_caller = 0;
 
-  printk("_PCf(%s):%p:%s:%s:%d\n",file->f_dentry->d_name.name,file,fname,fxn,line);
   BUG_ON(!file);
   dentry = file->f_dentry;
   sb = dentry->d_sb;
-- 
1.5.4.3

