From 08cf23733867b8599d0b969124aff7f618342f6c Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Tue, 5 Jun 2007 21:58:07 -0400
Subject: [PATCH] bugfix: properly reset lower inode start/end range

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/commonfops.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/commonfops.c b/fs/unionfs/commonfops.c
index 0e923df..900887e 100644
--- a/fs/unionfs/commonfops.c
+++ b/fs/unionfs/commonfops.c
@@ -723,7 +723,8 @@ static int unionfs_ioctl_queryfile(struct file *file, unsigned int cmd,
 	/* restore original dentry's offsets */
 	set_dbstart(dentry, orig_bstart);
 	set_dbend(dentry, orig_bend);
-	ibstart(dentry->d_inode) = ibend(dentry->d_inode) = orig_bend;
+	ibstart(dentry->d_inode) = orig_bstart;
+	ibend(dentry->d_inode) = orig_bend;
 
 	err = copy_to_user((void __user *)arg, &branchlist, sizeof(fd_set));
 	if (err)
-- 
1.5.4.3

