From 8032543892a1ec262770496837e0c006ba2f0074 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Fri, 29 Jun 2007 02:02:42 -0400
Subject: [PATCH] Unionfs: improved comment above unionfs_follow_link

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/inode.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 384800c..b2c0214 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -908,9 +908,9 @@ out:
  * unionfs_readlink *does* lock our dentry and revalidate the dentry.
  * Therefore, we do not have to lock our dentry here, to prevent a deadlock;
  * nor do we need to revalidate it either.  It is safe to not lock our
- * dentry here because unionfs_follow_link does not do anything (prior to
- * calling ->readlink) which could become inconsistent due to branch
- * management.
+ * dentry here, nor revalidate it, because unionfs_follow_link does not do
+ * anything (prior to calling ->readlink) which could become inconsistent
+ * due to branch management.
  */
 static void *unionfs_follow_link(struct dentry *dentry, struct nameidata *nd)
 {
-- 
1.5.4.3

