From ff16742f0106c68e307edaa84929855550abb4cd Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Fri, 13 Jul 2007 02:40:31 -0400
Subject: [PATCH] Unionfs: fix minor typo in a comment

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/super.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/super.c b/fs/unionfs/super.c
index 6c9bc1d..9b6f501 100644
--- a/fs/unionfs/super.c
+++ b/fs/unionfs/super.c
@@ -691,7 +691,7 @@ out_no_change:
 	 */
 
 	/*
-	 * No we call drop_pagecache_sb() to invalidate all pages in this
+	 * Now we call drop_pagecache_sb() to invalidate all pages in this
 	 * super.  This function calls invalidate_inode_pages(mapping),
 	 * which calls invalidate_mapping_pages(): the latter, however, will
 	 * not invalidate pages which are dirty, locked, under writeback, or
-- 
1.5.4.3

