From bf737fd22066a5f515d1e1ced4e2edba865df57a Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sun, 21 Oct 2007 16:10:59 -0400
Subject: [PATCH] Unionfs: remove obsolete #define and comment

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 include/linux/union_fs.h |    3 ---
 1 files changed, 0 insertions(+), 3 deletions(-)

diff --git a/include/linux/union_fs.h b/include/linux/union_fs.h
index d13eb48..a467de0 100644
--- a/include/linux/union_fs.h
+++ b/include/linux/union_fs.h
@@ -18,8 +18,5 @@
 # define UNIONFS_IOCTL_INCGEN		_IOR(0x15, 11, int)
 # define UNIONFS_IOCTL_QUERYFILE	_IOR(0x15, 15, int)
 
-/* We don't support normal remount, but unionctl uses it. */
-# define UNIONFS_REMOUNT_MAGIC		0x4a5a4380
-
 #endif /* _LINUX_UNIONFS_H */
 
-- 
1.5.4.3

