From b423fe96cd56cc84e200c9234835ae789b593db9 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sun, 27 Jul 2008 19:15:44 -0400
Subject: [PATCH] Unionfs: minor checkpatch fixes

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/main.c |   12 ------------
 1 files changed, 0 insertions(+), 12 deletions(-)

diff --git a/fs/unionfs/main.c b/fs/unionfs/main.c
index b5b6bd9..12fd13e 100644
--- a/fs/unionfs/main.c
+++ b/fs/unionfs/main.c
@@ -460,8 +460,6 @@ static struct unionfs_dentry_info *unionfs_parse_options(
 
 	while ((optname = strsep(&options, ",")) != NULL) {
 		char *optarg;
-		char *endptr;
-		int intval;
 
 		if (!optname || !*optname)
 			continue;
@@ -494,16 +492,6 @@ static struct unionfs_dentry_info *unionfs_parse_options(
 			continue;
 		}
 
-		/* All of these options require an integer argument. */
-		intval = simple_strtoul(optarg, &endptr, 0);
-		if (*endptr) {
-			printk(KERN_ERR
-			       "unionfs: invalid %s option '%s'\n",
-			       optname, optarg);
-			err = -EINVAL;
-			goto out_error;
-		}
-
 		err = -EINVAL;
 		printk(KERN_ERR
 		       "unionfs: unrecognized option '%s'\n", optname);
-- 
1.5.4.3

