--- src_gspi8686/app/wpa_supplicant-0.5.7/common.c.orig	2009-07-22 14:09:46.000000000 +0200
+++ src_gspi8686/app/wpa_supplicant-0.5.7/common.c	2009-07-22 14:10:17.000000000 +0200
@@ -617,7 +617,7 @@ const char * wpa_ssid_txt(u8 *ssid, size
 	os_memcpy(ssid_txt, ssid, ssid_len);
 	ssid_txt[ssid_len] = '\0';
 	for (pos = ssid_txt; *pos != '\0'; pos++) {
-		if ((u8) *pos < 32 || (u8) *pos >= 127)
+		if ((u8) *pos < 32)
 			*pos = '_';
 	}
 	return ssid_txt;
