Index: busybox_1_7_stable/networking/ifupdown.c
===================================================================
--- busybox_1_7_stable/networking/ifupdown.c	(revision 2550)
+++ busybox_1_7_stable/networking/ifupdown.c	(revision 2551)
@@ -151,6 +151,15 @@
 			return ifd->option[i].value;
 		}
 	}
+
+	/* gethostname if it is not in the interface definition */
+	if (strncmpz(id, "hostname", idlen) == 0) {
+		static char hostname[HOST_NAME_MAX + 1];
+		if (gethostname(hostname, HOST_NAME_MAX) == 0) {
+			return hostname;
+		}
+	}
+
 	return NULL;
 }
 
