--- busybox-1.13.4/networking/udhcp/dhcpc.c.orig	2009-08-28 15:50:57.000000000 +0200
+++ busybox-1.13.4/networking/udhcp/dhcpc.c	2009-08-28 15:51:33.000000000 +0200
@@ -287,9 +287,6 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 	/* Equivalent of doing a fflush after every \n */
 	setlinebuf(stdout);
 
-	/* Create pidfile */
-	write_pidfile(client_config.pidfile);
-
 	/* Goes to stdout (unless NOMMU) and possibly syslog */
 	bb_info_msg("%s (v"BB_VER") started", applet_name);
 
@@ -306,6 +303,9 @@ int udhcpc_main(int argc UNUSED_PARAM, c
 	/* setup the signal pipe */
 	udhcp_sp_setup();
 
+	/* Create pidfile */
+	write_pidfile(client_config.pidfile);
+
 	state = INIT_SELECTING;
 	udhcp_run_script(NULL, "deconfig");
 	change_listen_mode(LISTEN_RAW);
