Index: busybox_1_7_stable/networking/ifupdown.c
===================================================================
--- busybox_1_7_stable/networking/ifupdown.c	(revision 835)
+++ busybox_1_7_stable/networking/ifupdown.c	(revision 836)
@@ -505,7 +505,7 @@
 
 // -R was not in the file patched by Elvis. Check what it does
 	return execute("udhcpc -R -p /var/run/udhcpc.%iface%.pid "
-					"-b --zeroconf [[-s %script%]] "
+					"-b --syslog --zeroconf "
 			"-i %iface%[[ -H %hostname%]][[ -c %clientid%]][[ -s %script%]]",
 			ifd, exec);
 #else
