From 346419aa6dc059e95f1c320218ad6113f98385ec Mon Sep 17 00:00:00 2001
From: Adrian Brunyate <abrunyate@yahoo.com>
Date: Thu, 26 Apr 2007 18:51:16 -0400
Subject: [PATCH] Unionfs: Check remount options for being NULL

Signed-off-by: Adrian Brunyate <abrunyate@yahoo.com>
Signed-off-by: Josef 'Jeff' Sipek <jsipek@cs.sunysb.edu>
---
 fs/unionfs/super.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/super.c b/fs/unionfs/super.c
index ee12d03..02c0cc8 100644
--- a/fs/unionfs/super.c
+++ b/fs/unionfs/super.c
@@ -442,7 +442,7 @@ static int unionfs_remount_fs(struct super_block *sb, int *flags,
 	 * the union to a "ro" or "rw" and the VFS took care of it.  So
 	 * nothing to do and we're done.
 	 */
-	if (options[0] == '\0')
+	if (!options || options[0] == '\0')
 		goto out_error;
 
 	/*
-- 
1.5.4.3

