From 3ea2d08bc96fb04b183d05a63f89a3c7c695a0ab Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@bigvaio.(none)>
Date: Sun, 25 Mar 2007 01:12:48 -0400
Subject: [PATCH] cleanup: fix one odd indentation and long line breakup.

---
 fs/unionfs/inode.c |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 9042c0f..71dbf37 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -332,11 +332,10 @@ docopyup:
 		int bindex;
 
 		for (bindex = old_bstart - 1; bindex >= 0; bindex--) {
-			err =
-				copyup_dentry(old_dentry->d_parent->
-					      d_inode, old_dentry,
-					      old_bstart, bindex, NULL,
-					      old_dentry->d_inode->i_size);
+			err = copyup_dentry(old_dentry->d_parent->d_inode,
+					    old_dentry, old_bstart,
+					    bindex, NULL,
+					    old_dentry->d_inode->i_size);
 			if (!err) {
 				hidden_new_dentry =
 					create_parents(dir, new_dentry,
-- 
1.5.4.3

