From cbfbb13136ab68d7a73f370b67fd1deb4073a0bd Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@bigvaio.(none)>
Date: Fri, 18 May 2007 02:18:44 -0400
Subject: [PATCH] debugging: small fix to Makefile to correct #define typo

---
 fs/unionfs/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/Makefile b/fs/unionfs/Makefile
index 8c8e6e0..e9f8ad3 100644
--- a/fs/unionfs/Makefile
+++ b/fs/unionfs/Makefile
@@ -11,5 +11,5 @@ unionfs-$(CONFIG_UNION_FS_XATTR) += xattr.o
 
 ifeq ($(CONFIG_UNIONFS_DEBUG),y)
 unionfs-y += debug.o
-EXTRA_CFLAGS += -DUNIONFS_DDEBU=1
+EXTRA_CFLAGS += -DUNIONFS_DEBUG=1
 endif
-- 
1.5.4.3

