From 08f3bc52ca7303471c071625a3688accef2ba09b Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Sun, 22 Jul 2007 01:10:45 -0400
Subject: [PATCH] Unionfs: ensure cache coherency in unionfs_fsync/fasync

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/file.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/fs/unionfs/file.c b/fs/unionfs/file.c
index d08ef8b..e9e63b7 100644
--- a/fs/unionfs/file.c
+++ b/fs/unionfs/file.c
@@ -179,6 +179,8 @@ int unionfs_fsync(struct file *file, struct dentry *dentry, int datasync)
 			goto out;
 	}
 
+	unionfs_copy_attr_times(inode);
+
 out:
 	unionfs_read_unlock(file->f_path.dentry->d_sb);
 	unionfs_check_file(file);
@@ -222,6 +224,8 @@ int unionfs_fasync(int fd, struct file *file, int flag)
 			goto out;
 	}
 
+	unionfs_copy_attr_times(inode);
+
 out:
 	unionfs_read_unlock(file->f_path.dentry->d_sb);
 	unionfs_check_file(file);
-- 
1.5.4.3

