From 7c96c4178c6c8b51d75c82d71e0e299c85eb3a54 Mon Sep 17 00:00:00 2001
From: Erez_Zadok <ezk@cs.sunysb.edu>
Date: Thu, 6 Sep 2007 00:41:59 -0400
Subject: [PATCH] Unionfs: remove unnecessary comment

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/fanout.h |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/fanout.h b/fs/unionfs/fanout.h
index c5bf454..afeb9f6 100644
--- a/fs/unionfs/fanout.h
+++ b/fs/unionfs/fanout.h
@@ -320,7 +320,6 @@ static inline void unionfs_copy_attr_times(struct inode *upper)
 			upper->i_ctime = lower->i_ctime;
 		if (timespec_compare(&upper->i_atime, &lower->i_atime) < 0)
 			upper->i_atime = lower->i_atime;
-		/* XXX: should we notify_change on our upper inode? */
 	}
 }
 
-- 
1.5.4.3

