From ce3394e8234f00a8b375e72c7b5fbf608e72b099 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sat, 29 Sep 2007 01:36:57 -0400
Subject: [PATCH] Unionfs: properly indent static struct

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/main.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/main.c b/fs/unionfs/main.c
index 615617a..5c7f672 100644
--- a/fs/unionfs/main.c
+++ b/fs/unionfs/main.c
@@ -564,7 +564,10 @@ static struct dentry *unionfs_d_alloc_root(struct super_block *sb)
 	struct dentry *ret = NULL;
 
 	if (sb) {
-		static const struct qstr name = {.name = "/",.len = 1 };
+		static const struct qstr name = {
+			.name = "/",
+			.len = 1
+		};
 
 		ret = d_alloc(NULL, &name);
 		if (likely(ret)) {
-- 
1.5.4.3

