From 41970130c08d2babb28d1c9d872ee12c1697ca00 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Fri, 2 Nov 2007 19:56:49 -0400
Subject: [PATCH] Unionfs/VFS: no need to export 2 symbols in security/security.c

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 security/security.c |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/security/security.c b/security/security.c
index d349d55..59838a9 100644
--- a/security/security.c
+++ b/security/security.c
@@ -469,7 +469,6 @@ int security_inode_permission(struct inode *inode, int mask, struct nameidata *n
 		return 0;
 	return security_ops->inode_permission(inode, mask, nd);
 }
-EXPORT_SYMBOL(security_inode_permission);
 
 int security_inode_setattr(struct dentry *dentry, struct iattr *attr)
 {
@@ -584,7 +583,6 @@ int security_file_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 {
 	return security_ops->file_ioctl(file, cmd, arg);
 }
-EXPORT_SYMBOL(security_file_ioctl);
 
 int security_file_mmap(struct file *file, unsigned long reqprot,
 			unsigned long prot, unsigned long flags,
-- 
1.5.4.3

