From 4bfbac601e7718e5227973dd9f1fe985e815e831 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Tue, 20 Nov 2007 22:23:08 -0500
Subject: [PATCH] Unionfs: set lower mnt after mkdir which resulted in copyup

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/inode.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/inode.c b/fs/unionfs/inode.c
index 1708f40..9c144be 100644
--- a/fs/unionfs/inode.c
+++ b/fs/unionfs/inode.c
@@ -674,8 +674,10 @@ out:
 
 	kfree(name);
 
-	if (!err)
+	if (!err) {
 		unionfs_copy_attr_times(dentry->d_inode);
+		unionfs_postcopyup_setmnt(dentry);
+	}
 	unionfs_check_inode(parent);
 	unionfs_check_dentry(dentry);
 	unionfs_unlock_dentry(dentry);
-- 
1.5.4.3

