From f102ba82fad0d296fc47da8c9cd0231e59667888 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Fri, 28 Dec 2007 13:31:53 -0500
Subject: [PATCH] Unionfs: release 2.2

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/Makefile b/fs/unionfs/Makefile
index 6703a60..c37a765 100644
--- a/fs/unionfs/Makefile
+++ b/fs/unionfs/Makefile
@@ -1,4 +1,4 @@
-UNIONFS_VERSION="2.1.11 (for 2.6.24-rc4)"
+UNIONFS_VERSION="2.2 (for 2.6.24-rc6)"
 
 EXTRA_CFLAGS += -DUNIONFS_VERSION=\"$(UNIONFS_VERSION)\"
 
-- 
1.5.4.3

