From 4b7d24c39eff119dd7b9a0dbbe88adaeb17e5cd9 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sun, 27 Jul 2008 15:28:44 -0400
Subject: [PATCH] Unionfs: move fs/Makefile entry up to minimize merge conflicts

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/Makefile b/fs/Makefile
index 22ebb61..dadf53b 100644
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -85,6 +85,7 @@ obj-$(CONFIG_ISO9660_FS)	+= isofs/
 obj-$(CONFIG_HFSPLUS_FS)	+= hfsplus/ # Before hfs to find wrapped HFS+
 obj-$(CONFIG_HFS_FS)		+= hfs/
 obj-$(CONFIG_ECRYPT_FS)		+= ecryptfs/
+obj-$(CONFIG_UNION_FS)		+= unionfs/
 obj-$(CONFIG_VXFS_FS)		+= freevxfs/
 obj-$(CONFIG_NFS_FS)		+= nfs/
 obj-$(CONFIG_EXPORTFS)		+= exportfs/
@@ -119,4 +120,3 @@ obj-$(CONFIG_HPPFS)		+= hppfs/
 obj-$(CONFIG_DEBUG_FS)		+= debugfs/
 obj-$(CONFIG_OCFS2_FS)		+= ocfs2/
 obj-$(CONFIG_GFS2_FS)           += gfs2/
-obj-$(CONFIG_UNION_FS)		+= unionfs/
-- 
1.5.4.3

