From 2b18e837798b799d388d9e6bb6cdd480fb6e7287 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Thu, 18 Sep 2008 21:49:47 -0400
Subject: [PATCH] Unionfs: remove redundant file-invariant check in file_revalidate

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/commonfops.c |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/commonfops.c b/fs/unionfs/commonfops.c
index e68bc5a..214ad86 100644
--- a/fs/unionfs/commonfops.c
+++ b/fs/unionfs/commonfops.c
@@ -387,8 +387,6 @@ out:
 	if (err) {
 		kfree(UNIONFS_F(file)->lower_files);
 		kfree(UNIONFS_F(file)->saved_branch_ids);
-	} else {
-		unionfs_check_file(file);
 	}
 	return err;
 }
-- 
1.5.4.3

