From dcf28d3c59ed723606e78a5271df16471cb0e3c2 Mon Sep 17 00:00:00 2001
From: Erez Zadok <ezk@cs.sunysb.edu>
Date: Sun, 18 Jan 2009 21:22:52 -0500
Subject: [PATCH] Unionfs: release 2.5.1

Signed-off-by: Erez Zadok <ezk@cs.sunysb.edu>
---
 fs/unionfs/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/unionfs/Makefile b/fs/unionfs/Makefile
index 090915d..267fc9b 100644
--- a/fs/unionfs/Makefile
+++ b/fs/unionfs/Makefile
@@ -1,4 +1,4 @@
-UNIONFS_VERSION="2.5 (for 2.6.26.5)"
+UNIONFS_VERSION="2.5.1 (for 2.6.26.8)"
 
 EXTRA_CFLAGS += -DUNIONFS_VERSION=\"$(UNIONFS_VERSION)\"
 
-- 
1.5.4.3

