---
 drivers/net/fec.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.26/drivers/net/fec.c
===================================================================
--- linux-2.6.26.orig/drivers/net/fec.c	2009-04-28 11:15:43.000000000 +0100
+++ linux-2.6.26/drivers/net/fec.c	2009-04-28 11:16:23.000000000 +0100
@@ -2102,7 +2102,7 @@
 	clk_put(clk);
 
 	fep->phy_speed =
-		((((rate / 2 + 4999999) / 2500000) / 2) & 0x3F) << 1;
+		(((((rate / 2) + (4999999 / 2)) / 2500000)) & 0x3F) << 1;
 	fecp->fec_mii_speed = fep->phy_speed;
 	fec_restart(dev, 0);
 }
