Index: linux-2.6.26/drivers/video/mxc/mxcfb.c
===================================================================
--- linux-2.6.26.orig/drivers/video/mxc/mxcfb.c	2009-11-13 09:57:34.000000000 +0000
+++ linux-2.6.26/drivers/video/mxc/mxcfb.c	2009-11-13 10:09:26.000000000 +0000
@@ -488,6 +488,7 @@
 	case MXCFB_WAIT_FOR_VSYNC:
 		{
 #ifndef CONFIG_ARCH_MX3
+#ifndef CONFIG_MACH_FAB4
 			mxcfb_drv_data.vsync_flag = 0;
 			ipu_enable_irq(IPU_IRQ_SDC_DISP3_VSYNC);
 			if (!wait_event_interruptible_timeout
@@ -504,6 +505,7 @@
 				break;
 			}
 #endif
+#endif
 			break;
 		}
 #ifdef CONFIG_FB_MXC_TVOUT
Index: linux-2.6.26/drivers/mxc/ipu/ipu_common.c
===================================================================
--- linux-2.6.26.orig/drivers/mxc/ipu/ipu_common.c	2009-11-13 10:08:04.000000000 +0000
+++ linux-2.6.26/drivers/mxc/ipu/ipu_common.c	2009-11-13 10:08:22.000000000 +0000
@@ -640,9 +640,11 @@
 
 	if (bufNum == 0) {
 		/*Mark buffer 0 as ready. */
+		__raw_writel(0, IPU_CHA_BUF1_RDY);
 		__raw_writel(1UL << dma_chan, IPU_CHA_BUF0_RDY);
 	} else {
 		/*Mark buffer 1 as ready. */
+		__raw_writel(0, IPU_CHA_BUF0_RDY);
 		__raw_writel(1UL << dma_chan, IPU_CHA_BUF1_RDY);
 	}
 	return 0;
