Index: linux-2.6.26/drivers/mmc/host/mx_sdhci.c
===================================================================
--- linux-2.6.26.orig/drivers/mmc/host/mx_sdhci.c	2009-04-23 22:38:22.000000000 +0100
+++ linux-2.6.26/drivers/mmc/host/mx_sdhci.c	2009-04-23 22:38:29.000000000 +0100
@@ -1691,7 +1691,10 @@
 	}
 	host->detect_irq = platform_get_irq(pdev, 1);
 	if (!host->detect_irq) {
-		host->flags &= ~SDHCI_CD_PRESENT;
+		/* our wlan chips are on the main board, so assume the
+		 * device is present if no detect irq is given. */
+		//host->flags &= ~SDHCI_CD_PRESENT;
+		host->flags |= SDHCI_CD_PRESENT;
 		if ((pdev->id >= 0) && (pdev->id < MXC_SDHCI_NUM))
 			mxc_fix_chips[pdev->id] = chip;
 		goto no_detect_irq;
