Index: linux-2.6.26/sound/arm/mxc-alsa-spdif.c
===================================================================
--- linux-2.6.26.orig/sound/arm/mxc-alsa-spdif.c	2009-09-22 18:17:45.000000000 +0100
+++ linux-2.6.26/sound/arm/mxc-alsa-spdif.c	2009-10-02 14:40:05.000000000 +0100
@@ -1724,9 +1724,7 @@
 
 		mxc_spdif->s[SNDRV_PCM_STREAM_PLAYBACK].id = "spdif tx";
 		/* init tx channel status default value */
-		mxc_spdif_control.ch_status[0] =
-		    IEC958_AES0_CON_NOT_COPYRIGHT |
-		    IEC958_AES0_CON_EMPHASIS_5015;
+		mxc_spdif_control.ch_status[0] = IEC958_AES0_CON_NOT_COPYRIGHT;
 		mxc_spdif_control.ch_status[1] = IEC958_AES1_CON_DIGDIGCONV_ID;
 		mxc_spdif_control.ch_status[2] = 0x00;
 		mxc_spdif_control.ch_status[3] =
