Index: linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c
===================================================================
--- linux-2.6.26.orig/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-06-09 06:17:31.000000000 +0800
+++ linux-2.6.26/arch/arm/mach-mx25/mx25_baby_gpio.c	2009-06-11 05:19:27.000000000 +0800
@@ -48,8 +48,8 @@
 
 	mxc_iomux_set_pad(BABY_DAC_SDIN,  	    BABY_DAC_PAD_CFG);
 	mxc_iomux_set_pad(BABY_DAC_SDOUT, 	    BABY_DAC_PAD_CFG);
-	mxc_iomux_set_pad(BABY_DAC_SCLK,  	    BABY_DAC_PAD_CFG);
-	mxc_iomux_set_pad(BABY_DAC_LRCLK, 	    BABY_DAC_PAD_CFG);
+	mxc_iomux_set_pad(BABY_DAC_SCLK,  	    BABY_DAC_PAD_CFG |PAD_CTL_HYS_SCHMITZ);
+	mxc_iomux_set_pad(BABY_DAC_LRCLK, 	    BABY_DAC_PAD_CFG |PAD_CTL_HYS_SCHMITZ);
 	mxc_iomux_set_pad(BABY_DAC_PA_MUTE,         BABY_DAC_PAD_CFG);
 	mxc_iomux_set_pad(BABY_DAC_PA_SHUTDOWN,     BABY_DAC_PAD_CFG);
 	mxc_iomux_set_pad(BABY_DAC_HP_DETECT,       BABY_DAC_PAD_CFG);
@@ -80,20 +80,9 @@
 	mxc_set_gpio_direction(BABY_WLAN_POWER_DOWN_BAR, GPIO_DIRECTION_OUTPUT);
 	mxc_set_gpio_dataout  (BABY_WLAN_RESET_BAR,      1);
 	mxc_set_gpio_dataout  (BABY_WLAN_POWER_DOWN_BAR, 0);
-	msleep(1000);
-
-	/* Power up WLAN */
-	/* FIXME move this to wlan sdi driver */
-
-	mxc_set_gpio_dataout  (BABY_WLAN_RESET_BAR,      1);
-	mxc_set_gpio_dataout  (BABY_WLAN_POWER_DOWN_BAR, 0);
-	msleep(1000);
-	mxc_set_gpio_dataout  (BABY_WLAN_RESET_BAR,      1);
-	mxc_set_gpio_dataout  (BABY_WLAN_POWER_DOWN_BAR, 0);
-	msleep(1000);
+	msleep(10);
 	mxc_set_gpio_dataout  (BABY_WLAN_RESET_BAR,      1);
 	mxc_set_gpio_dataout  (BABY_WLAN_POWER_DOWN_BAR, 1);
-	msleep(1000);
 
 
 	/* MSP430 pins */
