Index: linux-2.6.26/drivers/mtd/mtd_blkdevs.c
===================================================================
--- linux-2.6.26.orig/drivers/mtd/mtd_blkdevs.c	2008-07-13 22:51:29.000000000 +0100
+++ linux-2.6.26/drivers/mtd/mtd_blkdevs.c	2009-03-20 17:08:34.000000000 +0000
@@ -276,6 +276,11 @@
 		snprintf(gd->disk_name, sizeof(gd->disk_name),
 			 "%s%d", tr->name, new->devnum);
 
+	/* HACK: rename the device to allow it to be mounted at boot time
+	 * using /dev/mtdblock:name */
+	snprintf(gd->disk_name, sizeof(gd->disk_name),
+		 "%s:%s", tr->name, new->mtd->name);
+
 	/* 2.5 has capacity in units of 512 bytes while still
 	   having BLOCK_SIZE_BITS set to 10. Just to keep us amused. */
 	set_capacity(gd, (new->size * tr->blksize) >> 9);
